// @visual-json/react/dist/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  Breadcrumbs: () => Breadcrumbs,
  ContextMenu: () => ContextMenu,
  DiffView: () => DiffView,
  FormView: () => FormView,
  JsonEditor: () => JsonEditor,
  SearchBar: () => SearchBar,
  StudioContext: () => StudioContext,
  TreeView: () => TreeView,
  VisualJson: () => VisualJson,
  getVisibleNodes: () => getVisibleNodes,
  useStudio: () => useStudio
});
module.exports = __toCommonJS(index_exports);

// src/json-editor.tsx
var import_react10 = require("react");

// src/visual-json.tsx
var import_react2 = require("react");
var import_core4 = require("@visual-json/core");

// ../../@internal/ui/dist/index.mjs
var import_core = require("@visual-json/core");
var import_core2 = require("@visual-json/core");
function getVisibleNodes(root, isExpanded) {
  const result = [];
  function walk(node) {
    result.push(node);
    if (isExpanded(node.id) && (node.type === "object" || node.type === "array")) {
      for (const child of node.children) {
        walk(child);
      }
    }
  }
  walk(root);
  return result;
}
var LABEL_FIELDS = ["name", "type", "title", "id", "label", "key"];
function getDisplayKey(node, state) {
  if (node.parentId === null) return "/";
  const parent = state.nodesById.get(node.parentId);
  if (parent?.type !== "array" || node.type !== "object") return node.key;
  for (const field of LABEL_FIELDS) {
    const child = node.children.find((c) => c.key === field);
    if (child?.value != null && child.value !== "") {
      return String(child.value);
    }
  }
  return node.key;
}
function collectAllIds(node) {
  const ids = [node.id];
  for (const child of node.children) {
    ids.push(...collectAllIds(child));
  }
  return ids;
}
var DEFAULT_CSS_VARS = {
  "--vj-bg": "#1e1e1e",
  "--vj-bg-panel": "#252526",
  "--vj-bg-hover": "#2a2d2e",
  "--vj-bg-selected": "#2a5a1e",
  "--vj-bg-selected-muted": "#2a2d2e",
  "--vj-bg-match": "#3a3520",
  "--vj-bg-match-active": "#51502b",
  "--vj-border": "#333333",
  "--vj-border-subtle": "#2a2a2a",
  "--vj-text": "#cccccc",
  "--vj-text-muted": "#888888",
  "--vj-text-dim": "#666666",
  "--vj-text-dimmer": "#555555",
  "--vj-string": "#ce9178",
  "--vj-number": "#b5cea8",
  "--vj-boolean": "#569cd6",
  "--vj-accent": "#007acc",
  "--vj-accent-muted": "#094771",
  "--vj-input-bg": "#3c3c3c",
  "--vj-input-border": "#555555",
  "--vj-error": "#f48771",
  "--vj-font": "monospace",
  "--vj-input-font-size": "13px"
};
var EMPTY_SET = Object.freeze(/* @__PURE__ */ new Set());
function sortByTreeOrder(root, ids) {
  const result = [];
  function walk(node) {
    if (ids.has(node.id)) result.push(node.id);
    for (const child of node.children) walk(child);
  }
  walk(root);
  return result;
}
function computeDrop(tree, drag) {
  const { draggedNodeIds, dropTargetNodeId, dropPosition } = drag;
  if (draggedNodeIds.size === 0 || !dropTargetNodeId || !dropPosition)
    return null;
  const targetNode = tree.nodesById.get(dropTargetNodeId);
  if (!targetNode || !targetNode.parentId) return null;
  for (const id of draggedNodeIds) {
    if ((0, import_core.isDescendant)(tree, dropTargetNodeId, id)) return null;
  }
  const targetParentId = targetNode.parentId;
  const targetParent = tree.nodesById.get(targetParentId);
  if (!targetParent) return null;
  const parentChildren = targetParent.children;
  const orderedDragIds = parentChildren.filter((c) => draggedNodeIds.has(c.id)).map((c) => c.id);
  const allSameParent = orderedDragIds.length === draggedNodeIds.size && [...draggedNodeIds].every((id) => {
    const n = tree.nodesById.get(id);
    return n?.parentId === targetParentId;
  });
  if (allSameParent) {
    return (0, import_core.reorderChildrenMulti)(
      tree,
      targetParentId,
      orderedDragIds,
      dropTargetNodeId,
      dropPosition
    );
  }
  const orderedIds = sortByTreeOrder(tree.root, draggedNodeIds);
  const draggedNodes = orderedIds.map((id) => tree.nodesById.get(id)).filter((n) => !!n && n.parentId !== null).map((n) => structuredClone(n));
  let newTree = tree;
  for (const id of [...orderedIds].reverse()) {
    if (newTree.nodesById.has(id)) {
      newTree = (0, import_core.removeNode)(newTree, id);
    }
  }
  const updatedTarget = newTree.nodesById.get(dropTargetNodeId);
  if (!updatedTarget || !updatedTarget.parentId) return null;
  const updatedParent = newTree.nodesById.get(updatedTarget.parentId);
  if (!updatedParent) return null;
  let insertIdx = updatedParent.children.findIndex(
    (c) => c.id === dropTargetNodeId
  );
  if (dropPosition === "after") insertIdx++;
  for (let i = 0; i < draggedNodes.length; i++) {
    newTree = (0, import_core.insertNode)(
      newTree,
      updatedParent.id,
      draggedNodes[i],
      insertIdx + i
    );
  }
  return newTree;
}
function setMultiDragImage(dataTransfer, count, rootEl) {
  const ghost = document.createElement("div");
  ghost.textContent = `${count} selected`;
  const root = rootEl ?? document.querySelector("[data-form-container], [role='tree']");
  const cs = root ? getComputedStyle(root) : null;
  const bg = cs?.getPropertyValue("--vj-bg-selected").trim() || DEFAULT_CSS_VARS["--vj-bg-selected"];
  const fg = cs?.getPropertyValue("--vj-text-selected").trim() || cs?.getPropertyValue("--vj-text").trim() || DEFAULT_CSS_VARS["--vj-text"];
  const font = cs?.getPropertyValue("--vj-font").trim() || DEFAULT_CSS_VARS["--vj-font"];
  ghost.style.cssText = [
    "position:fixed",
    "top:-1000px",
    "left:-1000px",
    "padding:4px 12px",
    `background:${bg}`,
    `color:${fg}`,
    `font-family:${font}`,
    "font-size:13px",
    "border-radius:4px",
    "white-space:nowrap",
    "pointer-events:none"
  ].join(";");
  document.body.appendChild(ghost);
  dataTransfer.setDragImage(ghost, 0, 14);
  requestAnimationFrame(() => ghost.remove());
}
var DIFF_COLORS = {
  added: { bg: "#1e3a1e", marker: "+", label: "#4ec94e" },
  removed: { bg: "#3a1e1e", marker: "-", label: "#f48771" },
  changed: { bg: "#3a3a1e", marker: "~", label: "#dcdcaa" }
};
function formatValue(value) {
  if (value === void 0) return "";
  if (value === null) return "null";
  if (typeof value === "string") return JSON.stringify(value);
  if (typeof value === "object") {
    const json = JSON.stringify(value, null, 2);
    if (json.length > 80) {
      return JSON.stringify(value).slice(0, 77) + "...";
    }
    return json;
  }
  return String(value);
}

// src/context.ts
var import_react = require("react");
var StudioContext = (0, import_react.createContext)(null);
function useStudio() {
  const ctx = (0, import_react.useContext)(StudioContext);
  if (!ctx) {
    throw new Error("useStudio must be used within a <VisualJson> provider");
  }
  return ctx;
}

// src/selection-utils.ts
var import_core3 = require("@visual-json/core");
function computeSelectAllIds(tree, focusedNodeId, currentlySelected) {
  if (!focusedNodeId) return null;
  let node = tree.nodesById.get(focusedNodeId);
  if (!node) return null;
  while (node) {
    const parent = node.parentId ? tree.nodesById.get(node.parentId) : void 0;
    const siblings = parent ? parent.children : [tree.root];
    const siblingIds = new Set(siblings.map((s) => s.id));
    const allSelected = siblings.every((s) => currentlySelected.has(s.id));
    if (!allSelected) {
      return siblingIds;
    }
    if (!parent || !parent.parentId) return siblingIds;
    node = parent;
  }
  return null;
}
function computeRangeIds(visibleNodes, anchorId, targetId) {
  const anchorIdx = visibleNodes.findIndex((n) => n.id === anchorId);
  const targetIdx = visibleNodes.findIndex((n) => n.id === targetId);
  if (anchorIdx === -1 || targetIdx === -1) return null;
  const start = Math.min(anchorIdx, targetIdx);
  const end = Math.max(anchorIdx, targetIdx);
  const ids = /* @__PURE__ */ new Set();
  for (let i = start; i <= end; i++) {
    ids.add(visibleNodes[i].id);
  }
  return ids;
}
function deleteSelectedNodes(tree, selectedIds, visibleNodes) {
  const idsToDelete = [...selectedIds].filter((id) => {
    const node = tree.nodesById.get(id);
    if (!node || node.parentId === null) return false;
    let cur = tree.nodesById.get(node.parentId);
    while (cur) {
      if (selectedIds.has(cur.id)) return false;
      cur = cur.parentId ? tree.nodesById.get(cur.parentId) : void 0;
    }
    return true;
  });
  if (idsToDelete.length === 0) return { newTree: tree, nextFocusId: null };
  const firstDeletedIdx = visibleNodes.findIndex((n) => selectedIds.has(n.id));
  let newTree = tree;
  for (const id of idsToDelete) {
    if (newTree.nodesById.has(id)) {
      newTree = (0, import_core3.removeNode)(newTree, id);
    }
  }
  let nextFocusId = null;
  for (let i = firstDeletedIdx; i < visibleNodes.length; i++) {
    const id = visibleNodes[i].id;
    if (!selectedIds.has(id) && newTree.nodesById.has(id)) {
      nextFocusId = id;
      break;
    }
  }
  if (!nextFocusId) {
    for (let i = firstDeletedIdx - 1; i >= 0; i--) {
      const id = visibleNodes[i].id;
      if (!selectedIds.has(id) && newTree.nodesById.has(id)) {
        nextFocusId = id;
        break;
      }
    }
  }
  return { newTree, nextFocusId };
}

// src/visual-json.tsx
var import_jsx_runtime = require("react/jsx-runtime");
function VisualJson({
  value,
  onChange,
  schema,
  children
}) {
  const [tree, setTreeState] = (0, import_react2.useState)(() => (0, import_core4.fromJson)(value));
  const [focusedNodeId, setFocusedNodeId] = (0, import_react2.useState)(null);
  const [selectedNodeIds, setSelectedNodeIdsState] = (0, import_react2.useState)(
    () => /* @__PURE__ */ new Set()
  );
  const selectedNodeIdsRef = (0, import_react2.useRef)(/* @__PURE__ */ new Set());
  const setSelectedNodeIds = (0, import_react2.useCallback)((ids) => {
    selectedNodeIdsRef.current = ids;
    setSelectedNodeIdsState(ids);
  }, []);
  const anchorNodeIdRef = (0, import_react2.useRef)(null);
  const [anchorNodeId, setAnchorNodeIdState] = (0, import_react2.useState)(null);
  const [drillDownNodeId, setDrillDownNodeId] = (0, import_react2.useState)(null);
  const [expandedNodeIds, setExpandedNodeIds] = (0, import_react2.useState)(
    () => /* @__PURE__ */ new Set([tree.root.id])
  );
  const setAnchorNodeId = (0, import_react2.useCallback)((id) => {
    anchorNodeIdRef.current = id;
    setAnchorNodeIdState(id);
  }, []);
  const focusSelectAndDrillDown = (0, import_react2.useCallback)(
    (nodeId) => {
      setFocusedNodeId(nodeId);
      setSelectedNodeIds(nodeId ? /* @__PURE__ */ new Set([nodeId]) : /* @__PURE__ */ new Set());
      setAnchorNodeId(nodeId);
      setDrillDownNodeId(nodeId);
    },
    [setSelectedNodeIds, setAnchorNodeId]
  );
  const visibleNodes = (0, import_react2.useMemo)(
    () => getVisibleNodes(tree.root, (id) => expandedNodeIds.has(id)),
    [tree.root, expandedNodeIds]
  );
  const visibleNodesOverrideRef = (0, import_react2.useRef)(null);
  const setVisibleNodesOverride = (0, import_react2.useCallback)((nodes) => {
    visibleNodesOverrideRef.current = nodes;
  }, []);
  const historyRef = (0, import_react2.useRef)(new import_core4.History());
  const isInternalChange = (0, import_react2.useRef)(false);
  const hasMounted = (0, import_react2.useRef)(false);
  const [canUndo, setCanUndo] = (0, import_react2.useState)(false);
  const [canRedo, setCanRedo] = (0, import_react2.useState)(false);
  const [searchQuery, setSearchQueryState] = (0, import_react2.useState)("");
  const [searchMatches, setSearchMatches] = (0, import_react2.useState)([]);
  const [searchMatchIndex, setSearchMatchIndex] = (0, import_react2.useState)(0);
  const [searchMatchNodeIds, setSearchMatchNodeIds] = (0, import_react2.useState)(
    /* @__PURE__ */ new Set()
  );
  (0, import_react2.useEffect)(() => {
    historyRef.current.push(tree);
    setCanUndo(historyRef.current.canUndo);
    setCanRedo(historyRef.current.canRedo);
  }, []);
  (0, import_react2.useEffect)(() => {
    if (!hasMounted.current) {
      hasMounted.current = true;
      return;
    }
    if (isInternalChange.current) {
      isInternalChange.current = false;
      return;
    }
    const newTree = (0, import_core4.fromJson)(value);
    setTreeState(newTree);
    setExpandedNodeIds(/* @__PURE__ */ new Set([newTree.root.id]));
    focusSelectAndDrillDown(null);
    historyRef.current = new import_core4.History();
    historyRef.current.push(newTree);
    setCanUndo(false);
    setCanRedo(false);
    setSearchQueryState("");
    setSearchMatches([]);
    setSearchMatchIndex(0);
    setSearchMatchNodeIds(/* @__PURE__ */ new Set());
  }, [value]);
  const setTree = (0, import_react2.useCallback)(
    (newTree) => {
      setTreeState(newTree);
      historyRef.current.push(newTree);
      setCanUndo(historyRef.current.canUndo);
      setCanRedo(historyRef.current.canRedo);
      isInternalChange.current = true;
      onChange?.((0, import_core4.toJson)(newTree.root));
    },
    [onChange]
  );
  const undo = (0, import_react2.useCallback)(() => {
    const prev = historyRef.current.undo();
    if (prev) {
      setTreeState(prev);
      setCanUndo(historyRef.current.canUndo);
      setCanRedo(historyRef.current.canRedo);
      isInternalChange.current = true;
      onChange?.((0, import_core4.toJson)(prev.root));
    }
  }, [onChange]);
  const redo = (0, import_react2.useCallback)(() => {
    const next = historyRef.current.redo();
    if (next) {
      setTreeState(next);
      setCanUndo(historyRef.current.canUndo);
      setCanRedo(historyRef.current.canRedo);
      isInternalChange.current = true;
      onChange?.((0, import_core4.toJson)(next.root));
    }
  }, [onChange]);
  (0, import_react2.useEffect)(() => {
    function handleKeyDown(e) {
      const mod = e.metaKey || e.ctrlKey;
      if (mod && e.key === "z" && !e.shiftKey) {
        e.preventDefault();
        undo();
      } else if (mod && e.key === "z" && e.shiftKey) {
        e.preventDefault();
        redo();
      } else if (mod && e.key === "y") {
        e.preventDefault();
        redo();
      }
    }
    document.addEventListener("keydown", handleKeyDown);
    return () => document.removeEventListener("keydown", handleKeyDown);
  }, [undo, redo]);
  const selectNode = (0, import_react2.useCallback)(
    (nodeId) => {
      setFocusedNodeId(nodeId);
      setSelectedNodeIds(nodeId ? /* @__PURE__ */ new Set([nodeId]) : /* @__PURE__ */ new Set());
      setAnchorNodeId(nodeId);
    },
    [setSelectedNodeIds, setAnchorNodeId]
  );
  const selectAndDrillDown = focusSelectAndDrillDown;
  const toggleNodeSelection = (0, import_react2.useCallback)(
    (nodeId) => {
      const next = new Set(selectedNodeIdsRef.current);
      if (next.has(nodeId)) {
        next.delete(nodeId);
      } else {
        next.add(nodeId);
      }
      setSelectedNodeIds(next);
      if (next.size === 0) {
        setFocusedNodeId(null);
        setAnchorNodeId(null);
      } else {
        setFocusedNodeId(nodeId);
        setAnchorNodeId(nodeId);
      }
    },
    [setSelectedNodeIds, setAnchorNodeId]
  );
  const selectNodeRange = (0, import_react2.useCallback)(
    (toNodeId) => {
      const nodes = visibleNodesOverrideRef.current ?? visibleNodes;
      const anchor = anchorNodeIdRef.current;
      if (!anchor) {
        setFocusedNodeId(toNodeId);
        setSelectedNodeIds(/* @__PURE__ */ new Set([toNodeId]));
        setAnchorNodeId(toNodeId);
        return;
      }
      const rangeIds = computeRangeIds(nodes, anchor, toNodeId);
      if (!rangeIds) {
        setFocusedNodeId(toNodeId);
        setSelectedNodeIds(/* @__PURE__ */ new Set([toNodeId]));
        setAnchorNodeId(toNodeId);
        return;
      }
      setSelectedNodeIds(rangeIds);
      setFocusedNodeId(toNodeId);
    },
    [visibleNodes, setSelectedNodeIds, setAnchorNodeId]
  );
  const setSelection = (0, import_react2.useCallback)(
    (focusedId, newSelectedIds, newAnchorId) => {
      setFocusedNodeId(focusedId);
      setSelectedNodeIds(newSelectedIds);
      setAnchorNodeId(newAnchorId);
    },
    [setSelectedNodeIds, setAnchorNodeId]
  );
  const drillDown = (0, import_react2.useCallback)((nodeId) => {
    setDrillDownNodeId(nodeId);
  }, []);
  const toggleExpand = (0, import_react2.useCallback)((nodeId) => {
    setExpandedNodeIds((prev) => {
      const next = new Set(prev);
      if (next.has(nodeId)) next.delete(nodeId);
      else next.add(nodeId);
      return next;
    });
  }, []);
  const expandNode = (0, import_react2.useCallback)((nodeId) => {
    setExpandedNodeIds((prev) => {
      const next = new Set(prev);
      next.add(nodeId);
      return next;
    });
  }, []);
  const collapseNode = (0, import_react2.useCallback)((nodeId) => {
    setExpandedNodeIds((prev) => {
      const next = new Set(prev);
      next.delete(nodeId);
      return next;
    });
  }, []);
  const expandAll = (0, import_react2.useCallback)(() => {
    const allIds = collectAllIds(tree.root);
    setExpandedNodeIds(new Set(allIds));
  }, [tree]);
  const collapseAll = (0, import_react2.useCallback)(() => {
    setExpandedNodeIds(/* @__PURE__ */ new Set([tree.root.id]));
  }, [tree]);
  const setSearchQuery = (0, import_react2.useCallback)(
    (query) => {
      setSearchQueryState(query);
      if (!query.trim()) {
        setSearchMatches([]);
        setSearchMatchIndex(0);
        setSearchMatchNodeIds(/* @__PURE__ */ new Set());
        return;
      }
      const matches = (0, import_core4.searchNodes)(tree, query);
      setSearchMatches(matches);
      setSearchMatchIndex(0);
      const matchIds = new Set(matches.map((m) => m.nodeId));
      setSearchMatchNodeIds(matchIds);
      if (matches.length > 0) {
        const firstId = matches[0].nodeId;
        const ancestors = (0, import_core4.getAncestorIds)(
          tree,
          matches.map((m) => m.nodeId)
        );
        setExpandedNodeIds((prev) => {
          const next = new Set(prev);
          for (const id of ancestors) next.add(id);
          return next;
        });
        focusSelectAndDrillDown(firstId);
      }
    },
    [tree]
  );
  const nextSearchMatch = (0, import_react2.useCallback)(() => {
    if (searchMatches.length === 0) return;
    const nextIdx = (searchMatchIndex + 1) % searchMatches.length;
    setSearchMatchIndex(nextIdx);
    focusSelectAndDrillDown(searchMatches[nextIdx].nodeId);
  }, [searchMatches, searchMatchIndex, focusSelectAndDrillDown]);
  const prevSearchMatch = (0, import_react2.useCallback)(() => {
    if (searchMatches.length === 0) return;
    const prevIdx = (searchMatchIndex - 1 + searchMatches.length) % searchMatches.length;
    setSearchMatchIndex(prevIdx);
    focusSelectAndDrillDown(searchMatches[prevIdx].nodeId);
  }, [searchMatches, searchMatchIndex, focusSelectAndDrillDown]);
  (0, import_react2.useEffect)(() => {
    if (!searchQuery.trim()) return;
    const matches = (0, import_core4.searchNodes)(tree, searchQuery);
    setSearchMatches(matches);
    setSearchMatchIndex(
      (prev) => Math.min(prev, Math.max(matches.length - 1, 0))
    );
    setSearchMatchNodeIds(new Set(matches.map((m) => m.nodeId)));
  }, [tree]);
  const state = (0, import_react2.useMemo)(
    () => ({
      tree,
      focusedNodeId,
      selectedNodeIds,
      anchorNodeId,
      drillDownNodeId,
      expandedNodeIds,
      schema: schema ?? null,
      searchQuery,
      searchMatches,
      searchMatchIndex,
      searchMatchNodeIds
    }),
    [
      tree,
      focusedNodeId,
      selectedNodeIds,
      anchorNodeId,
      drillDownNodeId,
      expandedNodeIds,
      schema,
      searchQuery,
      searchMatches,
      searchMatchIndex,
      searchMatchNodeIds
    ]
  );
  const actions = (0, import_react2.useMemo)(
    () => ({
      setTree,
      selectNode,
      selectAndDrillDown,
      toggleNodeSelection,
      selectNodeRange,
      setSelection,
      setVisibleNodesOverride,
      drillDown,
      toggleExpand,
      expandNode,
      collapseNode,
      expandAll,
      collapseAll,
      undo,
      redo,
      canUndo,
      canRedo,
      setSearchQuery,
      nextSearchMatch,
      prevSearchMatch
    }),
    [
      setTree,
      selectNode,
      selectAndDrillDown,
      toggleNodeSelection,
      selectNodeRange,
      setSelection,
      setVisibleNodesOverride,
      drillDown,
      toggleExpand,
      expandNode,
      collapseNode,
      expandAll,
      collapseAll,
      undo,
      redo,
      canUndo,
      canRedo,
      setSearchQuery,
      nextSearchMatch,
      prevSearchMatch
    ]
  );
  const contextValue = (0, import_react2.useMemo)(() => ({ state, actions }), [state, actions]);
  return /* @__PURE__ */ (0, import_jsx_runtime.jsx)(StudioContext.Provider, { value: contextValue, children });
}

// src/tree-view.tsx
var import_react5 = require("react");
var import_core6 = require("@visual-json/core");

// src/context-menu.tsx
var import_react3 = require("react");
var import_jsx_runtime2 = require("react/jsx-runtime");
function ContextMenu({ x, y, items, onClose }) {
  const ref = (0, import_react3.useRef)(null);
  const [pos, setPos] = (0, import_react3.useState)({ left: x, top: y });
  (0, import_react3.useEffect)(() => {
    if (!ref.current) return;
    const rect = ref.current.getBoundingClientRect();
    const vw = window.innerWidth;
    const vh = window.innerHeight;
    setPos({
      left: rect.right > vw ? Math.max(0, x - rect.width) : x,
      top: rect.bottom > vh ? Math.max(0, y - rect.height) : y
    });
  }, [x, y]);
  (0, import_react3.useEffect)(() => {
    function handleClick(e) {
      if (ref.current && !ref.current.contains(e.target)) {
        onClose();
      }
    }
    function handleKey(e) {
      if (e.key === "Escape") onClose();
    }
    document.addEventListener("mousedown", handleClick);
    document.addEventListener("keydown", handleKey);
    return () => {
      document.removeEventListener("mousedown", handleClick);
      document.removeEventListener("keydown", handleKey);
    };
  }, [onClose]);
  return /* @__PURE__ */ (0, import_jsx_runtime2.jsx)(
    "div",
    {
      ref,
      style: {
        position: "fixed",
        left: pos.left,
        top: pos.top,
        zIndex: 1e4,
        backgroundColor: "var(--vj-bg-panel, #252526)",
        border: "1px solid var(--vj-border, #454545)",
        borderRadius: 4,
        boxShadow: "0 4px 12px rgba(0,0,0,0.5)",
        padding: "4px 0",
        minWidth: 160
      },
      children: items.map(
        (item, i) => item.separator ? /* @__PURE__ */ (0, import_jsx_runtime2.jsx)(
          "div",
          {
            style: {
              height: 1,
              backgroundColor: "var(--vj-border, #454545)",
              margin: "4px 0"
            }
          },
          i
        ) : /* @__PURE__ */ (0, import_jsx_runtime2.jsx)(
          "button",
          {
            onClick: () => {
              if (!item.disabled) {
                item.action();
                onClose();
              }
            },
            disabled: item.disabled,
            style: {
              display: "block",
              width: "100%",
              textAlign: "left",
              background: "none",
              border: "none",
              color: item.disabled ? "var(--vj-text-dimmer, #555555)" : "var(--vj-text, #cccccc)",
              cursor: item.disabled ? "default" : "pointer",
              padding: "4px 16px",
              fontSize: 12,
              fontFamily: "var(--vj-font, monospace)"
            },
            onMouseEnter: (e) => {
              if (!item.disabled)
                e.target.style.backgroundColor = "var(--vj-accent-muted, #094771)";
            },
            onMouseLeave: (e) => {
              e.target.style.backgroundColor = "transparent";
            },
            children: item.label
          },
          i
        )
      )
    }
  );
}

// src/use-drag-drop.ts
var import_react4 = require("react");
var import_core5 = require("@visual-json/core");
var EMPTY_SET2 = Object.freeze(/* @__PURE__ */ new Set());
var INITIAL_DRAG_STATE = {
  draggedNodeIds: EMPTY_SET2,
  dropTargetNodeId: null,
  dropPosition: null
};
function setMultiDragImage2(e, count) {
  setMultiDragImage(e.dataTransfer, count);
}
function useDragDrop(visibleNodes, selectedNodeIds) {
  const { state, actions } = useStudio();
  const [dragState, setDragState] = (0, import_react4.useState)(INITIAL_DRAG_STATE);
  const dragStateRef = (0, import_react4.useRef)(dragState);
  dragStateRef.current = dragState;
  const visibleNodeIndexMap = (0, import_react4.useMemo)(() => {
    const map = /* @__PURE__ */ new Map();
    visibleNodes.forEach((n, i) => map.set(n.id, i));
    return map;
  }, [visibleNodes]);
  const handleDragStart = (0, import_react4.useCallback)(
    (nodeId) => {
      let ids;
      if (selectedNodeIds.size > 0 && selectedNodeIds.has(nodeId)) {
        ids = selectedNodeIds;
      } else {
        ids = /* @__PURE__ */ new Set([nodeId]);
      }
      setDragState({
        draggedNodeIds: ids,
        dropTargetNodeId: null,
        dropPosition: null
      });
    },
    [selectedNodeIds]
  );
  const rawDragOver = (0, import_react4.useCallback)(
    (nodeId, position) => {
      const draggedIds = dragStateRef.current.draggedNodeIds;
      for (const draggedId of draggedIds) {
        if (nodeId === draggedId || (0, import_core5.isDescendant)(state.tree, nodeId, draggedId)) {
          return;
        }
      }
      setDragState((prev) => ({
        ...prev,
        dropTargetNodeId: nodeId,
        dropPosition: position
      }));
    },
    [state.tree]
  );
  const handleDragOver = (0, import_react4.useCallback)(
    (nodeId, position) => {
      if (position === "before") {
        const idx = visibleNodeIndexMap.get(nodeId);
        if (idx !== void 0 && idx > 0) {
          rawDragOver(visibleNodes[idx - 1].id, "after");
          return;
        }
      }
      rawDragOver(nodeId, position);
    },
    [visibleNodes, visibleNodeIndexMap, rawDragOver]
  );
  const handleDragEnd = (0, import_react4.useCallback)(() => {
    setDragState(INITIAL_DRAG_STATE);
  }, []);
  const handleDrop = (0, import_react4.useCallback)(() => {
    const currentDragState = dragStateRef.current;
    const newTree = computeDrop(state.tree, currentDragState);
    if (newTree) {
      actions.setTree(newTree);
    }
    setDragState(INITIAL_DRAG_STATE);
  }, [state.tree, actions]);
  return {
    dragState,
    handleDragStart,
    handleDragOver,
    handleDragEnd,
    handleDrop
  };
}

// src/tree-view.tsx
var import_jsx_runtime3 = require("react/jsx-runtime");
function TreeNodeRow({
  node,
  depth,
  dragState,
  showValues,
  showCounts,
  isFocused,
  onSelectRange,
  onDragStart,
  onDragOver,
  onDragEnd,
  onDrop,
  onContextMenu
}) {
  const { state, actions } = useStudio();
  const isSelected = state.selectedNodeIds.has(node.id);
  const isExpanded = state.expandedNodeIds.has(node.id);
  const isContainer = node.type === "object" || node.type === "array";
  const [hovered, setHovered] = (0, import_react5.useState)(false);
  const isRoot = node.parentId === null;
  const isSearchMatch = state.searchMatchNodeIds.has(node.id);
  const isActiveMatch = state.searchMatches.length > 0 && state.searchMatches[state.searchMatchIndex]?.nodeId === node.id;
  const isDragTarget = dragState.dropTargetNodeId === node.id;
  const isDraggedNode = dragState.draggedNodeIds.has(node.id);
  function displayValue() {
    if (isContainer) {
      return node.type === "array" ? `[${node.children.length}]` : `{${node.children.length}}`;
    }
    if (node.value === null) return "null";
    if (typeof node.value === "string") return node.value;
    return String(node.value);
  }
  function handleDragOverEvent(e) {
    e.preventDefault();
    const rect = e.currentTarget.getBoundingClientRect();
    const midY = rect.top + rect.height / 2;
    const position = e.clientY < midY ? "before" : "after";
    onDragOver(node.id, position);
  }
  let borderTopColor = "transparent";
  let borderBottomColor = "transparent";
  if (isDragTarget && dragState.dropPosition === "before") {
    borderTopColor = "var(--vj-accent, #007acc)";
  } else if (isDragTarget && dragState.dropPosition === "after") {
    borderBottomColor = "var(--vj-accent, #007acc)";
  }
  return /* @__PURE__ */ (0, import_jsx_runtime3.jsxs)(import_jsx_runtime3.Fragment, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime3.jsxs)(
      "div",
      {
        role: "treeitem",
        "aria-selected": isSelected,
        "aria-expanded": isContainer ? isExpanded : void 0,
        onClick: (e) => {
          if (e.shiftKey) {
            onSelectRange(node.id);
          } else if (e.metaKey || e.ctrlKey) {
            actions.toggleNodeSelection(node.id);
          } else {
            actions.selectAndDrillDown(node.id);
          }
        },
        onMouseEnter: () => setHovered(true),
        onMouseLeave: () => setHovered(false),
        onContextMenu: (e) => onContextMenu(e, node),
        draggable: !isRoot,
        onDragStart: (e) => {
          e.dataTransfer.effectAllowed = "move";
          if (state.selectedNodeIds.size > 1 && state.selectedNodeIds.has(node.id)) {
            setMultiDragImage2(e, state.selectedNodeIds.size);
          }
          onDragStart(node.id);
        },
        onDragOver: handleDragOverEvent,
        onDragEnd,
        onDrop: (e) => {
          e.preventDefault();
          onDrop();
        },
        "data-node-id": node.id,
        style: {
          display: "flex",
          alignItems: "center",
          gap: 6,
          padding: "1px 8px",
          paddingLeft: 8 + depth * 16,
          cursor: "pointer",
          backgroundColor: isSelected ? isFocused ? "var(--vj-bg-selected, #2a5a1e)" : "var(--vj-bg-selected-muted, var(--vj-bg-hover, #2a2d2e))" : isActiveMatch ? "var(--vj-bg-match-active, #51502b)" : isSearchMatch ? "var(--vj-bg-match, #3a3520)" : hovered ? "var(--vj-bg-hover, #2a2d2e)" : "transparent",
          minHeight: 28,
          userSelect: "none",
          opacity: isDraggedNode ? 0.4 : 1,
          borderTop: `2px solid ${borderTopColor}`,
          borderBottom: `2px solid ${borderBottomColor}`,
          boxSizing: "border-box",
          color: isSelected && isFocused ? "var(--vj-text-selected, var(--vj-text, #cccccc))" : "var(--vj-text, #cccccc)"
        },
        children: [
          isContainer ? /* @__PURE__ */ (0, import_jsx_runtime3.jsx)(
            "button",
            {
              onClick: (e) => {
                e.stopPropagation();
                actions.toggleExpand(node.id);
              },
              "aria-label": isExpanded ? "Collapse" : "Expand",
              style: {
                background: "none",
                border: "none",
                color: "inherit",
                cursor: "pointer",
                padding: 0,
                width: 16,
                fontSize: 9,
                flexShrink: 0,
                display: "inline-flex",
                alignItems: "center",
                justifyContent: "center",
                transition: "transform 0.15s",
                transform: isExpanded ? "rotate(90deg)" : "rotate(0deg)"
              },
              children: "\u25B6"
            }
          ) : /* @__PURE__ */ (0, import_jsx_runtime3.jsx)("span", { style: { width: 16, flexShrink: 0 } }),
          /* @__PURE__ */ (0, import_jsx_runtime3.jsx)(
            "span",
            {
              style: {
                color: "inherit",
                fontSize: 13,
                fontFamily: "var(--vj-font, monospace)",
                flexShrink: 0,
                fontWeight: isRoot ? 600 : 400
              },
              children: isRoot ? "/" : getDisplayKey(node, state.tree)
            }
          ),
          !isRoot && isContainer && showCounts && /* @__PURE__ */ (0, import_jsx_runtime3.jsx)(
            "span",
            {
              style: {
                color: isSelected ? "inherit" : "var(--vj-text-muted, #888888)",
                fontSize: 13,
                fontFamily: "var(--vj-font, monospace)",
                whiteSpace: "nowrap",
                marginLeft: "auto"
              },
              children: displayValue()
            }
          ),
          !isRoot && !isContainer && showValues && /* @__PURE__ */ (0, import_jsx_runtime3.jsx)(
            "span",
            {
              style: {
                color: node.type === "string" ? "var(--vj-string, #ce9178)" : "var(--vj-number, #b5cea8)",
                fontSize: 13,
                fontFamily: "var(--vj-font, monospace)",
                overflow: "hidden",
                textOverflow: "ellipsis",
                whiteSpace: "nowrap",
                marginLeft: "auto"
              },
              children: displayValue()
            }
          )
        ]
      }
    ),
    isExpanded && node.children.map((child) => /* @__PURE__ */ (0, import_jsx_runtime3.jsx)(
      TreeNodeRow,
      {
        node: child,
        depth: depth + 1,
        dragState,
        showValues,
        showCounts,
        isFocused,
        onSelectRange,
        onDragStart,
        onDragOver,
        onDragEnd,
        onDrop,
        onContextMenu
      },
      child.id
    ))
  ] });
}
function TreeView({
  className,
  showValues = true,
  showCounts = false
}) {
  const { state, actions } = useStudio();
  const containerRef = (0, import_react5.useRef)(null);
  const visibleNodes = (0, import_react5.useMemo)(
    () => getVisibleNodes(state.tree.root, (id) => state.expandedNodeIds.has(id)),
    [state.tree.root, state.expandedNodeIds]
  );
  const {
    dragState,
    handleDragStart,
    handleDragOver,
    handleDragEnd,
    handleDrop
  } = useDragDrop(visibleNodes, state.selectedNodeIds);
  const handleSelectRange = (0, import_react5.useCallback)(
    (nodeId) => {
      actions.setVisibleNodesOverride(visibleNodes);
      actions.selectNodeRange(nodeId);
    },
    [visibleNodes, actions]
  );
  const [contextMenu, setContextMenu] = (0, import_react5.useState)(null);
  const handleContextMenu = (0, import_react5.useCallback)(
    (e, node) => {
      e.preventDefault();
      if (!state.selectedNodeIds.has(node.id)) {
        actions.selectAndDrillDown(node.id);
      }
      setContextMenu({ x: e.clientX, y: e.clientY, node });
    },
    [actions, state.selectedNodeIds]
  );
  const buildContextMenuItems = (0, import_react5.useCallback)(
    (node) => {
      const items = [];
      const isContainer = node.type === "object" || node.type === "array";
      if (isContainer) {
        items.push({
          label: "Expand all children",
          action: () => {
            function collectIds(n) {
              const ids = [n.id];
              for (const c of n.children) ids.push(...collectIds(c));
              return ids;
            }
            const allIds = collectIds(node);
            for (const id of allIds) actions.expandNode(id);
          }
        });
        items.push({
          label: "Collapse all children",
          action: () => {
            function collectIds(n) {
              const ids = [];
              for (const c of n.children) {
                ids.push(c.id);
                ids.push(...collectIds(c));
              }
              return ids;
            }
            const childIds = collectIds(node);
            for (const id of childIds) actions.collapseNode(id);
          }
        });
        items.push({ separator: true });
      }
      items.push({
        label: "Copy path",
        action: () => navigator.clipboard.writeText(node.path).catch(() => {
        })
      });
      items.push({
        label: "Copy value as JSON",
        action: () => {
          const val = (0, import_core6.toJson)(node);
          const text = typeof val === "string" ? val : JSON.stringify(val, null, 2);
          navigator.clipboard.writeText(text).catch(() => {
          });
        }
      });
      if (node.parentId) {
        items.push({ separator: true });
        items.push({
          label: "Duplicate",
          action: () => {
            const newTree = (0, import_core6.duplicateNode)(state.tree, node.id);
            actions.setTree(newTree);
          }
        });
        const typeSubmenu = [
          "string",
          "number",
          "boolean",
          "null",
          "object",
          "array"
        ].filter((t) => t !== node.type).map((t) => ({
          label: `Change to ${t}`,
          action: () => {
            const newTree = (0, import_core6.changeType)(state.tree, node.id, t);
            actions.setTree(newTree);
          }
        }));
        items.push({ separator: true });
        items.push(...typeSubmenu);
        items.push({ separator: true });
        items.push({
          label: "Delete",
          action: () => {
            const newTree = (0, import_core6.removeNode)(state.tree, node.id);
            actions.setTree(newTree);
          }
        });
      }
      return items;
    },
    [state.tree, actions]
  );
  const handleKeyDown = (0, import_react5.useCallback)(
    (e) => {
      const currentIndex = visibleNodes.findIndex(
        (n) => n.id === state.focusedNodeId
      );
      switch (e.key) {
        case "ArrowDown": {
          e.preventDefault();
          const next = visibleNodes[currentIndex + 1];
          if (next) {
            if (e.shiftKey) {
              handleSelectRange(next.id);
            } else {
              actions.selectNode(next.id);
            }
          }
          break;
        }
        case "ArrowUp": {
          e.preventDefault();
          const prev = visibleNodes[currentIndex - 1];
          if (prev) {
            if (e.shiftKey) {
              handleSelectRange(prev.id);
            } else {
              actions.selectNode(prev.id);
            }
          }
          break;
        }
        case "ArrowRight": {
          e.preventDefault();
          const node = currentIndex >= 0 ? visibleNodes[currentIndex] : null;
          if (node && (node.type === "object" || node.type === "array")) {
            if (!state.expandedNodeIds.has(node.id)) {
              actions.expandNode(node.id);
            } else if (node.children.length > 0) {
              actions.selectNode(node.children[0].id);
            }
          }
          break;
        }
        case "ArrowLeft": {
          e.preventDefault();
          const current = currentIndex >= 0 ? visibleNodes[currentIndex] : null;
          if (!current) break;
          const isContainer = current.type === "object" || current.type === "array";
          if (isContainer && state.expandedNodeIds.has(current.id)) {
            actions.collapseNode(current.id);
          } else if (current.parentId) {
            actions.selectNode(current.parentId);
          }
          break;
        }
        case "a": {
          if (e.metaKey || e.ctrlKey) {
            e.preventDefault();
            const ids = computeSelectAllIds(
              state.tree,
              state.focusedNodeId,
              state.selectedNodeIds
            );
            if (ids) {
              actions.setSelection(
                state.focusedNodeId,
                ids,
                state.focusedNodeId
              );
            }
          }
          break;
        }
        case "Escape": {
          e.preventDefault();
          if (state.selectedNodeIds.size > 1 && state.focusedNodeId) {
            actions.selectNode(state.focusedNodeId);
          } else {
            actions.setSelection(null, /* @__PURE__ */ new Set(), null);
          }
          break;
        }
        case "Delete":
        case "Backspace": {
          e.preventDefault();
          const { newTree, nextFocusId } = deleteSelectedNodes(
            state.tree,
            state.selectedNodeIds,
            visibleNodes
          );
          if (newTree === state.tree) break;
          actions.setTree(newTree);
          if (nextFocusId) {
            actions.selectNode(nextFocusId);
          } else {
            actions.setSelection(null, /* @__PURE__ */ new Set(), null);
          }
          break;
        }
      }
    },
    [
      visibleNodes,
      state.focusedNodeId,
      state.selectedNodeIds,
      state.expandedNodeIds,
      state.tree,
      actions,
      handleSelectRange
    ]
  );
  const [isFocused, setIsFocused] = (0, import_react5.useState)(false);
  (0, import_react5.useEffect)(() => {
    if (state.focusedNodeId && containerRef.current) {
      const el = containerRef.current.querySelector(
        `[data-node-id="${state.focusedNodeId}"]`
      );
      if (el) {
        el.scrollIntoView({ block: "nearest" });
      }
    }
  }, [state.focusedNodeId]);
  return /* @__PURE__ */ (0, import_jsx_runtime3.jsxs)(import_jsx_runtime3.Fragment, { children: [
    /* @__PURE__ */ (0, import_jsx_runtime3.jsx)(
      "div",
      {
        ref: containerRef,
        className,
        role: "tree",
        tabIndex: 0,
        onKeyDown: handleKeyDown,
        onFocus: () => setIsFocused(true),
        onBlur: () => setIsFocused(false),
        style: {
          overflow: "auto",
          backgroundColor: "var(--vj-bg, #1e1e1e)",
          color: "var(--vj-text, #cccccc)",
          fontFamily: "var(--vj-font, monospace)",
          fontSize: 13,
          outline: "none",
          flex: 1
        },
        children: /* @__PURE__ */ (0, import_jsx_runtime3.jsx)(
          TreeNodeRow,
          {
            node: state.tree.root,
            depth: 0,
            dragState,
            showValues,
            showCounts,
            isFocused,
            onSelectRange: handleSelectRange,
            onDragStart: handleDragStart,
            onDragOver: handleDragOver,
            onDragEnd: handleDragEnd,
            onDrop: handleDrop,
            onContextMenu: handleContextMenu
          }
        )
      }
    ),
    contextMenu && /* @__PURE__ */ (0, import_jsx_runtime3.jsx)(
      ContextMenu,
      {
        x: contextMenu.x,
        y: contextMenu.y,
        items: buildContextMenuItems(contextMenu.node),
        onClose: () => setContextMenu(null)
      }
    )
  ] });
}

// src/form-view.tsx
var import_react8 = require("react");
var import_core7 = require("@visual-json/core");

// src/breadcrumbs.tsx
var import_react6 = require("react");
var import_jsx_runtime4 = require("react/jsx-runtime");
var MAX_SUGGESTIONS = 20;
var DROPDOWN_MAX_HEIGHT = 200;
function Breadcrumbs({ className }) {
  const { state, actions } = useStudio();
  const drillDownNode = state.drillDownNodeId ? state.tree.nodesById.get(state.drillDownNodeId) : null;
  const currentPath = drillDownNode?.path ?? "/";
  const [inputValue, setInputValue] = (0, import_react6.useState)(currentPath);
  const [open, setOpen] = (0, import_react6.useState)(false);
  const [highlightIndex, setHighlightIndex] = (0, import_react6.useState)(0);
  const inputRef = (0, import_react6.useRef)(null);
  const listRef = (0, import_react6.useRef)(null);
  const wrapperRef = (0, import_react6.useRef)(null);
  (0, import_react6.useEffect)(() => {
    setInputValue(currentPath);
  }, [currentPath]);
  const suggestions = (0, import_react6.useMemo)(() => {
    if (!open) return [];
    const query = inputValue.toLowerCase();
    const matches = [];
    for (const [id, node] of state.tree.nodesById) {
      if (node.path.toLowerCase().startsWith(query)) {
        matches.push({ id, path: node.path });
      }
      if (matches.length >= MAX_SUGGESTIONS) break;
    }
    matches.sort((a, b) => a.path.localeCompare(b.path));
    return matches;
  }, [open, inputValue, state.tree.nodesById]);
  (0, import_react6.useEffect)(() => {
    setHighlightIndex(0);
  }, [suggestions]);
  const navigateTo = (0, import_react6.useCallback)(
    (path) => {
      for (const [id, node] of state.tree.nodesById) {
        if (node.path === path) {
          actions.selectAndDrillDown(id);
          break;
        }
      }
      setOpen(false);
      inputRef.current?.blur();
    },
    [state.tree.nodesById, actions]
  );
  const handleKeyDown = (0, import_react6.useCallback)(
    (e) => {
      if (!open) {
        if (e.key === "ArrowDown" || e.key === "Enter") {
          setOpen(true);
          e.preventDefault();
        }
        return;
      }
      switch (e.key) {
        case "ArrowDown":
          e.preventDefault();
          setHighlightIndex((i) => Math.min(i + 1, suggestions.length - 1));
          break;
        case "ArrowUp":
          e.preventDefault();
          setHighlightIndex((i) => Math.max(i - 1, 0));
          break;
        case "Enter":
          e.preventDefault();
          if (suggestions.length > 0 && highlightIndex < suggestions.length) {
            navigateTo(suggestions[highlightIndex].path);
          } else {
            const trimmed = inputValue.trim() || "/";
            navigateTo(trimmed);
          }
          break;
        case "Escape":
          e.preventDefault();
          setInputValue(currentPath);
          setOpen(false);
          inputRef.current?.blur();
          break;
      }
    },
    [open, suggestions, highlightIndex, inputValue, currentPath, navigateTo]
  );
  (0, import_react6.useEffect)(() => {
    const el = listRef.current;
    if (!el || !open) return;
    const item = el.children[highlightIndex];
    item?.scrollIntoView({ block: "nearest" });
  }, [highlightIndex, open]);
  (0, import_react6.useEffect)(() => {
    function handleClickOutside(e) {
      if (wrapperRef.current && !wrapperRef.current.contains(e.target)) {
        setInputValue(currentPath);
        setOpen(false);
      }
    }
    document.addEventListener("mousedown", handleClickOutside);
    return () => document.removeEventListener("mousedown", handleClickOutside);
  }, [currentPath]);
  return /* @__PURE__ */ (0, import_jsx_runtime4.jsxs)(
    "div",
    {
      ref: wrapperRef,
      className,
      style: { position: "relative", flex: 1, minWidth: 0 },
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime4.jsx)(
          "input",
          {
            ref: inputRef,
            value: inputValue,
            onChange: (e) => {
              setInputValue(e.target.value);
              if (!open) setOpen(true);
            },
            onFocus: (e) => {
              e.target.select();
              setOpen(true);
            },
            onKeyDown: handleKeyDown,
            spellCheck: false,
            autoComplete: "off",
            style: {
              width: "100%",
              boxSizing: "border-box",
              padding: "3px 0",
              fontSize: "var(--vj-input-font-size, 13px)",
              fontFamily: "var(--vj-font, monospace)",
              color: "var(--vj-text-muted, #999999)",
              background: "transparent",
              border: "none",
              outline: "none"
            }
          }
        ),
        open && suggestions.length > 0 && /* @__PURE__ */ (0, import_jsx_runtime4.jsx)(
          "div",
          {
            ref: listRef,
            style: {
              position: "absolute",
              top: "100%",
              left: -12,
              right: -12,
              zIndex: 50,
              maxHeight: DROPDOWN_MAX_HEIGHT,
              overflowY: "auto",
              backgroundColor: "var(--vj-bg-panel, #252526)",
              border: "1px solid var(--vj-border, #333333)",
              boxShadow: "0 4px 12px rgba(0,0,0,0.3)"
            },
            children: suggestions.map((s, i) => /* @__PURE__ */ (0, import_jsx_runtime4.jsx)(
              "div",
              {
                onMouseDown: (e) => {
                  e.preventDefault();
                  navigateTo(s.path);
                },
                onMouseEnter: () => setHighlightIndex(i),
                style: {
                  padding: "4px 12px",
                  fontSize: 13,
                  fontFamily: "var(--vj-font, monospace)",
                  color: i === highlightIndex ? "var(--vj-text, #cccccc)" : "var(--vj-text-muted, #888888)",
                  backgroundColor: i === highlightIndex ? "var(--vj-bg-hover, #2a2d2e)" : "transparent",
                  cursor: "pointer",
                  whiteSpace: "nowrap",
                  overflow: "hidden",
                  textOverflow: "ellipsis"
                },
                children: s.path
              },
              s.id
            ))
          }
        )
      ]
    }
  );
}

// src/enum-input.tsx
var import_react7 = require("react");
var import_jsx_runtime5 = require("react/jsx-runtime");
var DROPDOWN_MAX_HEIGHT2 = 200;
function EnumInput({
  enumValues,
  value,
  onValueChange,
  inputRef,
  inputStyle
}) {
  const [inputValue, setInputValue] = (0, import_react7.useState)(value);
  const [open, setOpen] = (0, import_react7.useState)(false);
  const [highlightIndex, setHighlightIndex] = (0, import_react7.useState)(0);
  const listRef = (0, import_react7.useRef)(null);
  const wrapperRef = (0, import_react7.useRef)(null);
  (0, import_react7.useEffect)(() => {
    setInputValue(value);
  }, [value]);
  const suggestions = (0, import_react7.useMemo)(
    () => enumValues.map((v) => String(v)),
    [enumValues]
  );
  (0, import_react7.useEffect)(() => {
    setHighlightIndex(0);
  }, [suggestions]);
  const selectValue = (0, import_react7.useCallback)(
    (val) => {
      onValueChange(val);
      setInputValue(val);
      setOpen(false);
    },
    [onValueChange]
  );
  const handleKeyDown = (0, import_react7.useCallback)(
    (e) => {
      if (!open) {
        if (e.key === "ArrowDown" || e.key === "ArrowUp") {
          e.preventDefault();
          e.stopPropagation();
          setOpen(true);
        }
        return;
      }
      switch (e.key) {
        case "ArrowDown":
          e.preventDefault();
          e.stopPropagation();
          setHighlightIndex((i) => Math.min(i + 1, suggestions.length - 1));
          break;
        case "ArrowUp":
          e.preventDefault();
          e.stopPropagation();
          setHighlightIndex((i) => Math.max(i - 1, 0));
          break;
        case "Enter":
          e.preventDefault();
          e.stopPropagation();
          if (suggestions.length > 0 && highlightIndex < suggestions.length) {
            selectValue(suggestions[highlightIndex]);
          }
          break;
        case "Escape":
          e.preventDefault();
          e.stopPropagation();
          setInputValue(value);
          setOpen(false);
          break;
        case "Tab":
          setInputValue(value);
          setOpen(false);
          break;
      }
    },
    [open, suggestions, highlightIndex, value, selectValue]
  );
  (0, import_react7.useEffect)(() => {
    const el = listRef.current;
    if (!el || !open) return;
    const item = el.children[highlightIndex];
    item?.scrollIntoView({ block: "nearest" });
  }, [highlightIndex, open]);
  (0, import_react7.useEffect)(() => {
    function handleClickOutside(e) {
      if (wrapperRef.current && !wrapperRef.current.contains(e.target)) {
        setInputValue(value);
        setOpen(false);
      }
    }
    document.addEventListener("mousedown", handleClickOutside);
    return () => document.removeEventListener("mousedown", handleClickOutside);
  }, [value]);
  return /* @__PURE__ */ (0, import_jsx_runtime5.jsxs)(
    "div",
    {
      ref: wrapperRef,
      style: { position: "relative", flex: 1, minWidth: 0 },
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime5.jsx)(
          "input",
          {
            ref: inputRef,
            value: inputValue,
            onChange: (e) => {
              setInputValue(e.target.value);
              if (!open) setOpen(true);
            },
            onFocus: () => setOpen(true),
            onKeyDown: handleKeyDown,
            onClick: (e) => e.stopPropagation(),
            spellCheck: false,
            autoComplete: "off",
            style: inputStyle
          }
        ),
        open && suggestions.length > 0 && /* @__PURE__ */ (0, import_jsx_runtime5.jsx)(
          "div",
          {
            ref: listRef,
            style: {
              position: "absolute",
              top: "calc(100% + 4px)",
              left: -32,
              right: 0,
              zIndex: 50,
              maxHeight: DROPDOWN_MAX_HEIGHT2,
              overflowY: "auto",
              backgroundColor: "var(--vj-bg-panel, #252526)",
              border: "1px solid var(--vj-border, #333333)",
              boxShadow: "0 4px 12px rgba(0,0,0,0.3)"
            },
            children: suggestions.map((s, i) => /* @__PURE__ */ (0, import_jsx_runtime5.jsxs)(
              "div",
              {
                onMouseDown: (e) => {
                  e.preventDefault();
                  selectValue(s);
                },
                onMouseEnter: () => setHighlightIndex(i),
                style: {
                  padding: "4px 12px",
                  fontSize: 13,
                  fontFamily: "var(--vj-font, monospace)",
                  display: "flex",
                  alignItems: "center",
                  gap: 6,
                  color: i === highlightIndex ? "var(--vj-text, #cccccc)" : "var(--vj-text-muted, #888888)",
                  backgroundColor: i === highlightIndex ? "var(--vj-bg-hover, #2a2d2e)" : "transparent",
                  cursor: "pointer",
                  whiteSpace: "nowrap",
                  overflow: "hidden",
                  textOverflow: "ellipsis"
                },
                children: [
                  /* @__PURE__ */ (0, import_jsx_runtime5.jsx)("span", { style: { width: 14, flexShrink: 0, fontSize: 12 }, children: s === value ? "\u2713" : "" }),
                  s
                ]
              },
              s
            ))
          }
        )
      ]
    }
  );
}

// src/form-view.tsx
var import_jsx_runtime6 = require("react/jsx-runtime");
function getResolvedSchema(schema, rootSchema, path) {
  if (!schema) return void 0;
  const raw = (0, import_core7.getPropertySchema)(schema, path, rootSchema);
  if (!raw) return void 0;
  return (0, import_core7.resolveRef)(raw, rootSchema ?? schema);
}
function getValueColor(node) {
  if (node.type === "boolean" || node.type === "null")
    return "var(--vj-boolean, #569cd6)";
  if (node.type === "number") return "var(--vj-number, #b5cea8)";
  return "var(--vj-string, #ce9178)";
}
function getDisplayValue(node) {
  if (node.type === "null") return "null";
  if (node.type === "boolean") return String(node.value);
  if (node.value === null || node.value === void 0) return "";
  return String(node.value);
}
function FormField({
  node,
  schema,
  rootSchema,
  depth,
  showDescriptions,
  showCounts,
  editingNodeId,
  collapsedIds,
  maxKeyLength,
  maxDepth,
  isFocused,
  dragState,
  onSelect,
  onToggleCollapse,
  onStartEditing,
  onDragStart,
  onDragOver,
  onDragEnd,
  onDrop
}) {
  const { state, actions } = useStudio();
  const isContainer = node.type === "object" || node.type === "array";
  const collapsed = collapsedIds.has(node.id);
  const isSelected = state.selectedNodeIds.has(node.id);
  const isEditing = editingNodeId === node.id;
  const propSchema = getResolvedSchema(schema, rootSchema, node.path);
  const isRequired = checkRequired(node, schema, rootSchema);
  const [hovered, setHovered] = (0, import_react8.useState)(false);
  const isRoot = node.parentId === null;
  const isDragTarget = dragState.dropTargetNodeId === node.id;
  const isDraggedNode = dragState.draggedNodeIds.has(node.id);
  function handleDragOverEvent(e) {
    e.preventDefault();
    const rect = e.currentTarget.getBoundingClientRect();
    const midY = rect.top + rect.height / 2;
    onDragOver(node.id, e.clientY < midY ? "before" : "after");
  }
  let borderTopColor = "transparent";
  let borderBottomColor = "transparent";
  if (isDragTarget && dragState.dropPosition === "before") {
    borderTopColor = "var(--vj-accent, #007acc)";
  } else if (isDragTarget && dragState.dropPosition === "after") {
    borderBottomColor = "var(--vj-accent, #007acc)";
  }
  const valueRef = (0, import_react8.useRef)(null);
  const keyRef = (0, import_react8.useRef)(null);
  (0, import_react8.useEffect)(() => {
    if (!isEditing) return;
    if (!isContainer) {
      const hasValue = node.value !== null && node.value !== void 0 && node.value !== "";
      if (hasValue && valueRef.current) {
        valueRef.current.focus();
      } else if (keyRef.current) {
        keyRef.current.focus();
      }
    } else if (keyRef.current) {
      keyRef.current.focus();
    }
  }, [isEditing, isContainer, node.value]);
  const handleValueChange = (0, import_react8.useCallback)(
    (newValue) => {
      let parsed;
      if (propSchema?.type === "boolean" || newValue === "true" || newValue === "false") {
        parsed = newValue === "true";
      } else if (newValue === "null") {
        parsed = null;
      } else if (propSchema?.type === "number" || propSchema?.type === "integer" || node.type === "number") {
        const num = Number(newValue);
        parsed = isNaN(num) ? newValue : num;
      } else {
        parsed = newValue;
      }
      const newTree = (0, import_core7.setValue)(state.tree, node.id, parsed);
      actions.setTree(newTree);
    },
    [state.tree, node.id, node.type, actions, propSchema]
  );
  const handleKeyChange = (0, import_react8.useCallback)(
    (newKey) => {
      const newTree = (0, import_core7.setKey)(state.tree, node.id, newKey);
      actions.setTree(newTree);
    },
    [state.tree, node.id, actions]
  );
  const handleRemove = (0, import_react8.useCallback)(() => {
    const newTree = (0, import_core7.removeNode)(state.tree, node.id);
    actions.setTree(newTree);
  }, [state.tree, node.id, actions]);
  const handleAddChild = (0, import_react8.useCallback)(() => {
    const key = node.type === "array" ? String(node.children.length) : `newKey${node.children.length}`;
    const newTree = (0, import_core7.addProperty)(state.tree, node.id, key, "");
    actions.setTree(newTree);
  }, [state.tree, node.id, node.type, node.children.length, actions]);
  const description = propSchema?.description;
  const isDeprecated = propSchema?.deprecated;
  const fieldTitle = propSchema?.title;
  const parentIsObject = node.parentId && state.tree.nodesById.get(node.parentId)?.type === "object";
  const rowBg = isSelected ? isFocused ? "var(--vj-bg-selected, #2a5a1e)" : "var(--vj-bg-selected-muted, var(--vj-bg-hover, #2a2d2e))" : hovered ? "var(--vj-bg-hover, #2a2d2e)" : "transparent";
  const rowColor = isSelected && isFocused ? "var(--vj-text-selected, var(--vj-text, #cccccc))" : "var(--vj-text, #cccccc)";
  if (isContainer) {
    return /* @__PURE__ */ (0, import_jsx_runtime6.jsxs)("div", { children: [
      /* @__PURE__ */ (0, import_jsx_runtime6.jsxs)(
        "div",
        {
          "data-form-node-id": node.id,
          draggable: !isRoot,
          onDragStart: (e) => {
            e.dataTransfer.effectAllowed = "move";
            if (state.selectedNodeIds.size > 1 && state.selectedNodeIds.has(node.id)) {
              setMultiDragImage2(e, state.selectedNodeIds.size);
            }
            onDragStart(node.id);
          },
          onDragOver: handleDragOverEvent,
          onDragEnd,
          onDrop: (e) => {
            e.preventDefault();
            onDrop();
          },
          style: {
            display: "flex",
            alignItems: "center",
            gap: 6,
            padding: "1px 8px",
            paddingLeft: 8 + depth * 16,
            cursor: "pointer",
            backgroundColor: rowBg,
            color: rowColor,
            height: 28,
            boxSizing: "border-box",
            userSelect: "none",
            opacity: isDeprecated ? 0.5 : isDraggedNode ? 0.4 : 1,
            borderTop: `2px solid ${borderTopColor}`,
            borderBottom: `2px solid ${borderBottomColor}`
          },
          onClick: (e) => {
            e.stopPropagation();
            onSelect(node.id, e);
          },
          onDoubleClick: () => onToggleCollapse(node.id),
          onMouseEnter: () => setHovered(true),
          onMouseLeave: () => setHovered(false),
          children: [
            /* @__PURE__ */ (0, import_jsx_runtime6.jsx)(
              "button",
              {
                onClick: (e) => {
                  e.stopPropagation();
                  onToggleCollapse(node.id);
                },
                style: {
                  background: "none",
                  border: "none",
                  color: "inherit",
                  cursor: "pointer",
                  padding: 0,
                  width: 16,
                  fontSize: 9,
                  flexShrink: 0,
                  display: "inline-flex",
                  alignItems: "center",
                  justifyContent: "center",
                  transition: "transform 0.15s",
                  transform: collapsed ? "rotate(0deg)" : "rotate(90deg)"
                },
                children: "\u25B6"
              }
            ),
            isEditing && !isRoot ? /* @__PURE__ */ (0, import_jsx_runtime6.jsx)(
              "input",
              {
                ref: keyRef,
                value: node.key,
                onChange: (e) => handleKeyChange(e.target.value),
                onClick: (e) => e.stopPropagation(),
                style: {
                  background: "none",
                  border: "none",
                  color: "inherit",
                  fontFamily: "var(--vj-font, monospace)",
                  fontSize: "var(--vj-input-font-size, 13px)",
                  fontWeight: 500,
                  padding: 0,
                  outline: "none",
                  flexShrink: 0,
                  width: `calc(${(maxDepth - depth) * 16}px + ${maxKeyLength}ch)`
                }
              }
            ) : /* @__PURE__ */ (0, import_jsx_runtime6.jsx)(
              "span",
              {
                style: {
                  color: !isRoot && !parentIsObject && !isSelected ? "var(--vj-text-muted, #888888)" : "inherit",
                  fontSize: "var(--vj-input-font-size, 13px)",
                  fontFamily: "var(--vj-font, monospace)",
                  fontWeight: 500,
                  flexShrink: 0,
                  display: "inline-block",
                  width: `calc(${(maxDepth - depth) * 16}px + ${maxKeyLength}ch)`
                },
                children: isRoot ? "/" : getDisplayKey(node, state.tree)
              }
            ),
            showDescriptions && fieldTitle && !isSelected && /* @__PURE__ */ (0, import_jsx_runtime6.jsx)(
              "span",
              {
                style: {
                  color: "var(--vj-text-muted, #888888)",
                  fontSize: 11,
                  fontFamily: "var(--vj-font, monospace)"
                },
                children: fieldTitle
              }
            ),
            hovered && /* @__PURE__ */ (0, import_jsx_runtime6.jsxs)(
              "button",
              {
                onClick: (e) => {
                  e.stopPropagation();
                  handleAddChild();
                },
                style: {
                  background: "none",
                  border: "none",
                  color: isSelected ? "inherit" : "var(--vj-text-muted, #888888)",
                  cursor: "pointer",
                  padding: 0,
                  fontSize: 12,
                  fontFamily: "var(--vj-font, monospace)"
                },
                children: [
                  "+ Add ",
                  node.type === "array" ? "item" : "property"
                ]
              }
            ),
            showCounts && /* @__PURE__ */ (0, import_jsx_runtime6.jsx)(
              "span",
              {
                style: {
                  color: isSelected ? "inherit" : "var(--vj-text-dim, #666666)",
                  fontSize: 12,
                  fontFamily: "var(--vj-font, monospace)",
                  marginLeft: "auto"
                },
                children: node.type === "array" ? `${node.children.length} items` : `${node.children.length} properties`
              }
            ),
            !isRoot && isEditing && /* @__PURE__ */ (0, import_jsx_runtime6.jsx)(
              "button",
              {
                onClick: (e) => {
                  e.stopPropagation();
                  handleRemove();
                },
                style: {
                  background: "none",
                  border: "none",
                  color: isSelected ? "inherit" : "var(--vj-text-muted, #888888)",
                  cursor: "pointer",
                  padding: "2px 4px",
                  fontSize: 14,
                  fontFamily: "var(--vj-font, monospace)",
                  ...!showCounts ? { marginLeft: "auto" } : {}
                },
                title: "Remove",
                children: "\xD7"
              }
            )
          ]
        }
      ),
      showDescriptions && description && /* @__PURE__ */ (0, import_jsx_runtime6.jsx)(
        "div",
        {
          style: {
            padding: "2px 12px 4px",
            paddingLeft: 8 + depth * 16 + 22,
            fontSize: 11,
            color: "var(--vj-text-dim, #666666)",
            fontFamily: "var(--vj-font, monospace)"
          },
          children: description
        }
      ),
      !collapsed && /* @__PURE__ */ (0, import_jsx_runtime6.jsx)("div", { children: node.children.map((child) => /* @__PURE__ */ (0, import_jsx_runtime6.jsx)(
        FormField,
        {
          node: child,
          schema,
          rootSchema,
          depth: depth + 1,
          showDescriptions,
          showCounts,
          editingNodeId,
          collapsedIds,
          maxKeyLength,
          maxDepth,
          isFocused,
          dragState,
          onSelect,
          onToggleCollapse,
          onStartEditing,
          onDragStart,
          onDragOver,
          onDragEnd,
          onDrop
        },
        child.id
      )) })
    ] });
  }
  const displayValue = getDisplayValue(node);
  const valueColor = getValueColor(node);
  return /* @__PURE__ */ (0, import_jsx_runtime6.jsxs)(
    "div",
    {
      "data-form-node-id": node.id,
      draggable: !isRoot,
      onDragStart: (e) => {
        e.dataTransfer.effectAllowed = "move";
        if (state.selectedNodeIds.size > 1 && state.selectedNodeIds.has(node.id)) {
          setMultiDragImage2(e, state.selectedNodeIds.size);
        }
        onDragStart(node.id);
      },
      onDragOver: handleDragOverEvent,
      onDragEnd,
      onDrop: (e) => {
        e.preventDefault();
        onDrop();
      },
      style: {
        display: "flex",
        alignItems: "center",
        gap: 6,
        padding: "1px 8px",
        paddingLeft: 8 + depth * 16,
        cursor: "pointer",
        backgroundColor: rowBg,
        color: rowColor,
        height: 28,
        boxSizing: "border-box",
        userSelect: "none",
        opacity: isDeprecated ? 0.5 : isDraggedNode ? 0.4 : 1,
        borderTop: `2px solid ${borderTopColor}`,
        borderBottom: `2px solid ${borderBottomColor}`
      },
      onClick: (e) => {
        e.stopPropagation();
        onSelect(node.id, e);
      },
      onDoubleClick: () => onStartEditing(node.id),
      onMouseEnter: () => setHovered(true),
      onMouseLeave: () => setHovered(false),
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime6.jsx)("span", { style: { width: 16, flexShrink: 0 } }),
        isEditing && parentIsObject ? /* @__PURE__ */ (0, import_jsx_runtime6.jsx)(
          "input",
          {
            ref: keyRef,
            value: node.key,
            onChange: (e) => handleKeyChange(e.target.value),
            onClick: (e) => e.stopPropagation(),
            onKeyDown: (e) => {
              if (e.key === "Tab" && !e.shiftKey && valueRef.current) {
                e.preventDefault();
                valueRef.current.focus();
              }
            },
            style: {
              background: "none",
              border: "none",
              color: "inherit",
              fontFamily: "var(--vj-font, monospace)",
              fontSize: "var(--vj-input-font-size, 13px)",
              padding: 0,
              flexShrink: 0,
              outline: "none",
              width: `calc(${(maxDepth - depth) * 16}px + ${maxKeyLength}ch)`
            }
          }
        ) : /* @__PURE__ */ (0, import_jsx_runtime6.jsx)(
          "span",
          {
            style: {
              color: !parentIsObject && !isSelected ? "var(--vj-text-muted, #888888)" : "inherit",
              fontSize: "var(--vj-input-font-size, 13px)",
              fontFamily: "var(--vj-font, monospace)",
              flexShrink: 0,
              display: "inline-block",
              width: `calc(${(maxDepth - depth) * 16}px + ${maxKeyLength}ch)`
            },
            children: getDisplayKey(node, state.tree)
          }
        ),
        isRequired && !isSelected && /* @__PURE__ */ (0, import_jsx_runtime6.jsx)(
          "span",
          {
            style: {
              color: "var(--vj-error, #f48771)",
              fontSize: 10,
              fontFamily: "var(--vj-font, monospace)"
            },
            children: "*"
          }
        ),
        isEditing ? /* @__PURE__ */ (0, import_jsx_runtime6.jsx)(
          "div",
          {
            style: {
              flex: 1,
              display: "flex",
              alignItems: "center",
              gap: 6,
              minWidth: 0
            },
            children: renderEditInput(
              node,
              propSchema,
              displayValue,
              handleValueChange,
              valueRef,
              valueColor
            )
          }
        ) : /* @__PURE__ */ (0, import_jsx_runtime6.jsx)(
          "span",
          {
            style: {
              color: valueColor,
              fontSize: "var(--vj-input-font-size, 13px)",
              fontFamily: "var(--vj-font, monospace)",
              overflow: "hidden",
              textOverflow: "ellipsis",
              whiteSpace: "nowrap",
              fontStyle: node.type === "null" ? "italic" : void 0
            },
            children: displayValue
          }
        ),
        isEditing && /* @__PURE__ */ (0, import_jsx_runtime6.jsx)(
          "button",
          {
            onClick: (e) => {
              e.stopPropagation();
              handleRemove();
            },
            style: {
              background: "none",
              border: "none",
              color: isSelected ? "inherit" : "var(--vj-text-muted, #888888)",
              cursor: "pointer",
              padding: "2px 4px",
              fontSize: 14,
              fontFamily: "var(--vj-font, monospace)",
              flexShrink: 0
            },
            title: "Remove",
            children: "\xD7"
          }
        )
      ]
    }
  );
}
function renderEditInput(node, propSchema, displayValue, handleValueChange, inputRef, valueColor) {
  const hasEnumValues = propSchema?.enum && propSchema.enum.length > 0;
  const inputStyle = {
    background: "none",
    border: "none",
    fontFamily: "var(--vj-font, monospace)",
    fontSize: "var(--vj-input-font-size, 13px)",
    padding: 0,
    flex: 1,
    outline: "none",
    color: valueColor
  };
  if (node.type === "boolean") {
    return /* @__PURE__ */ (0, import_jsx_runtime6.jsx)(
      EnumInput,
      {
        enumValues: ["true", "false"],
        value: String(node.value),
        onValueChange: handleValueChange,
        inputRef,
        inputStyle
      }
    );
  }
  if (hasEnumValues && propSchema?.enum) {
    return /* @__PURE__ */ (0, import_jsx_runtime6.jsx)(
      EnumInput,
      {
        enumValues: propSchema.enum,
        value: displayValue,
        onValueChange: handleValueChange,
        inputRef,
        inputStyle
      }
    );
  }
  if (node.type === "null") {
    return /* @__PURE__ */ (0, import_jsx_runtime6.jsx)(
      "span",
      {
        style: {
          color: "var(--vj-boolean, #569cd6)",
          fontFamily: "var(--vj-font, monospace)",
          fontSize: "var(--vj-input-font-size, 13px)",
          fontStyle: "italic",
          flex: 1
        },
        children: "null"
      }
    );
  }
  return /* @__PURE__ */ (0, import_jsx_runtime6.jsx)(
    "input",
    {
      ref: inputRef,
      value: displayValue,
      onChange: (e) => handleValueChange(e.target.value),
      onClick: (e) => e.stopPropagation(),
      placeholder: propSchema?.default !== void 0 ? String(propSchema.default) : "<value>",
      style: inputStyle
    }
  );
}
function checkRequired(node, schema, rootSchema) {
  if (!schema || !node.parentId) return false;
  const parentPath = node.path.split("/").slice(0, -1).join("/") || "/";
  const parentSchema = getResolvedSchema(schema, rootSchema, parentPath);
  return parentSchema?.required?.includes(node.key) ?? false;
}
function FormView({
  className,
  showDescriptions = false,
  showCounts = false
}) {
  const { state, actions } = useStudio();
  const rootSchema = state.schema ?? void 0;
  const drillDownNode = state.drillDownNodeId ? state.tree.nodesById.get(state.drillDownNodeId) : null;
  const displayNode = drillDownNode ?? state.tree.root;
  const [editingNodeId, setEditingNodeId] = (0, import_react8.useState)(null);
  const preEditTreeRef = (0, import_react8.useRef)(null);
  const [collapsedIds, setCollapsedIds] = (0, import_react8.useState)(
    () => /* @__PURE__ */ new Set()
  );
  const containerRef = (0, import_react8.useRef)(null);
  const [isFocused, setIsFocused] = (0, import_react8.useState)(false);
  (0, import_react8.useEffect)(() => {
    setEditingNodeId(null);
    setCollapsedIds(/* @__PURE__ */ new Set());
  }, [displayNode.id]);
  const visibleNodes = (0, import_react8.useMemo)(
    () => getVisibleNodes(displayNode, (id) => !collapsedIds.has(id)),
    [displayNode, collapsedIds]
  );
  const {
    dragState,
    handleDragStart,
    handleDragOver,
    handleDragEnd,
    handleDrop
  } = useDragDrop(visibleNodes, state.selectedNodeIds);
  (0, import_react8.useEffect)(() => {
    actions.setVisibleNodesOverride(visibleNodes);
    return () => actions.setVisibleNodesOverride(null);
  }, [visibleNodes, actions]);
  const { maxKeyLength, maxDepth } = (0, import_react8.useMemo)(() => {
    let maxKey = 1;
    let maxD = 0;
    const baseSegments = displayNode.path === "/" ? 0 : displayNode.path.split("/").filter(Boolean).length;
    for (const node of visibleNodes) {
      const keyText = node.parentId === null ? "/" : getDisplayKey(node, state.tree);
      if (keyText.length > maxKey) maxKey = keyText.length;
      const segments = node.path === "/" ? 0 : node.path.split("/").filter(Boolean).length;
      const depth = segments - baseSegments;
      if (depth > maxD) maxD = depth;
    }
    return { maxKeyLength: maxKey, maxDepth: maxD };
  }, [visibleNodes, displayNode.path, state.tree]);
  const handleSelect = (0, import_react8.useCallback)(
    (nodeId, e) => {
      setEditingNodeId(null);
      if (e.shiftKey) {
        actions.setVisibleNodesOverride(visibleNodes);
        actions.selectNodeRange(nodeId);
      } else if (e.metaKey || e.ctrlKey) {
        actions.toggleNodeSelection(nodeId);
      } else {
        actions.selectNode(nodeId);
      }
    },
    [actions, visibleNodes]
  );
  const handleToggleCollapse = (0, import_react8.useCallback)((nodeId) => {
    setCollapsedIds((prev) => {
      const next = new Set(prev);
      if (next.has(nodeId)) {
        next.delete(nodeId);
      } else {
        next.add(nodeId);
      }
      return next;
    });
  }, []);
  const handleStartEditing = (0, import_react8.useCallback)(
    (nodeId) => {
      preEditTreeRef.current = state.tree;
      setEditingNodeId(nodeId);
    },
    [state.tree]
  );
  const scrollToNode = (0, import_react8.useCallback)((nodeId) => {
    requestAnimationFrame(() => {
      const el = containerRef.current?.querySelector(
        `[data-form-node-id="${nodeId}"]`
      );
      el?.scrollIntoView({ block: "nearest" });
    });
  }, []);
  const handleKeyDown = (0, import_react8.useCallback)(
    (e) => {
      if (editingNodeId) {
        if (e.key === "Escape") {
          e.preventDefault();
          e.stopPropagation();
          if (preEditTreeRef.current) {
            actions.setTree(preEditTreeRef.current);
            preEditTreeRef.current = null;
          }
          setEditingNodeId(null);
          containerRef.current?.focus();
        } else if (e.key === "Enter") {
          e.preventDefault();
          e.stopPropagation();
          preEditTreeRef.current = null;
          setEditingNodeId(null);
          containerRef.current?.focus();
        }
        return;
      }
      let currentIndex = visibleNodes.findIndex(
        (n) => n.id === state.focusedNodeId
      );
      if (currentIndex === -1 && visibleNodes.length > 0) {
        currentIndex = 0;
      }
      switch (e.key) {
        case "ArrowDown": {
          e.preventDefault();
          const next = visibleNodes[currentIndex + 1];
          if (next) {
            if (e.shiftKey) {
              actions.setVisibleNodesOverride(visibleNodes);
              actions.selectNodeRange(next.id);
            } else {
              actions.selectNode(next.id);
            }
            scrollToNode(next.id);
          }
          break;
        }
        case "ArrowUp": {
          e.preventDefault();
          const prev = visibleNodes[currentIndex - 1];
          if (prev) {
            if (e.shiftKey) {
              actions.setVisibleNodesOverride(visibleNodes);
              actions.selectNodeRange(prev.id);
            } else {
              actions.selectNode(prev.id);
            }
            scrollToNode(prev.id);
          }
          break;
        }
        case "ArrowRight": {
          e.preventDefault();
          const node = currentIndex >= 0 ? visibleNodes[currentIndex] : null;
          if (node && (node.type === "object" || node.type === "array")) {
            if (collapsedIds.has(node.id)) {
              setCollapsedIds((prev) => {
                const next = new Set(prev);
                next.delete(node.id);
                return next;
              });
            } else if (node.children.length > 0) {
              actions.selectNode(node.children[0].id);
              scrollToNode(node.children[0].id);
            }
          }
          break;
        }
        case "ArrowLeft": {
          e.preventDefault();
          const current = currentIndex >= 0 ? visibleNodes[currentIndex] : null;
          if (!current) break;
          const isContainer = current.type === "object" || current.type === "array";
          if (isContainer && !collapsedIds.has(current.id)) {
            setCollapsedIds((prev) => {
              const next = new Set(prev);
              next.add(current.id);
              return next;
            });
          } else if (current.parentId) {
            const parentInVisible = visibleNodes.find(
              (n) => n.id === current.parentId
            );
            if (parentInVisible) {
              actions.selectNode(parentInVisible.id);
              scrollToNode(parentInVisible.id);
            }
          }
          break;
        }
        case "Enter": {
          e.preventDefault();
          if (state.focusedNodeId) {
            preEditTreeRef.current = state.tree;
            actions.selectNode(state.focusedNodeId);
            setEditingNodeId(state.focusedNodeId);
          }
          break;
        }
        case "a": {
          if (e.metaKey || e.ctrlKey) {
            e.preventDefault();
            const ids = computeSelectAllIds(
              state.tree,
              state.focusedNodeId,
              state.selectedNodeIds
            );
            if (ids) {
              actions.setSelection(
                state.focusedNodeId,
                ids,
                state.focusedNodeId
              );
            }
          }
          break;
        }
        case "Escape": {
          e.preventDefault();
          if (state.selectedNodeIds.size > 1 && state.focusedNodeId) {
            actions.selectNode(state.focusedNodeId);
          } else {
            actions.setSelection(null, /* @__PURE__ */ new Set(), null);
          }
          break;
        }
        case "Delete":
        case "Backspace": {
          e.preventDefault();
          const { newTree, nextFocusId } = deleteSelectedNodes(
            state.tree,
            state.selectedNodeIds,
            visibleNodes
          );
          if (newTree === state.tree) break;
          actions.setTree(newTree);
          if (nextFocusId) {
            actions.selectNode(nextFocusId);
          } else {
            actions.setSelection(null, /* @__PURE__ */ new Set(), null);
          }
          break;
        }
      }
    },
    [
      visibleNodes,
      state.focusedNodeId,
      state.selectedNodeIds,
      editingNodeId,
      collapsedIds,
      scrollToNode,
      state.tree,
      actions
    ]
  );
  return /* @__PURE__ */ (0, import_jsx_runtime6.jsxs)(
    "div",
    {
      className,
      style: {
        backgroundColor: "var(--vj-bg, #1e1e1e)",
        color: "var(--vj-text, #cccccc)",
        height: "100%",
        fontFamily: "var(--vj-font, monospace)",
        display: "flex",
        flexDirection: "column"
      },
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime6.jsx)(
          "div",
          {
            style: {
              display: "flex",
              alignItems: "center",
              padding: "4px 8px",
              borderBottom: "1px solid var(--vj-border, #333333)",
              backgroundColor: "var(--vj-bg, #1e1e1e)",
              flexShrink: 0
            },
            children: /* @__PURE__ */ (0, import_jsx_runtime6.jsx)(Breadcrumbs, {})
          }
        ),
        /* @__PURE__ */ (0, import_jsx_runtime6.jsx)(
          "div",
          {
            ref: containerRef,
            "data-form-container": true,
            tabIndex: 0,
            onKeyDown: handleKeyDown,
            onFocus: () => setIsFocused(true),
            onBlur: (e) => {
              if (!e.currentTarget.contains(e.relatedTarget)) {
                setIsFocused(false);
              }
            },
            style: {
              flex: 1,
              overflow: "auto",
              outline: "none"
            },
            children: /* @__PURE__ */ (0, import_jsx_runtime6.jsx)(
              FormField,
              {
                node: displayNode,
                schema: state.schema,
                rootSchema,
                depth: 0,
                showDescriptions,
                showCounts,
                editingNodeId,
                collapsedIds,
                maxKeyLength,
                maxDepth,
                isFocused,
                dragState,
                onSelect: handleSelect,
                onToggleCollapse: handleToggleCollapse,
                onStartEditing: handleStartEditing,
                onDragStart: handleDragStart,
                onDragOver: handleDragOver,
                onDragEnd: handleDragEnd,
                onDrop: handleDrop
              }
            )
          }
        )
      ]
    }
  );
}

// src/search-bar.tsx
var import_react9 = require("react");
var import_jsx_runtime7 = require("react/jsx-runtime");
function SearchBar({ className }) {
  const { state, actions } = useStudio();
  const inputRef = (0, import_react9.useRef)(null);
  const handleKeyDown = (0, import_react9.useCallback)(
    (e) => {
      if (e.key === "Enter") {
        e.preventDefault();
        if (e.shiftKey) {
          actions.prevSearchMatch();
        } else {
          actions.nextSearchMatch();
        }
      }
      if (e.key === "Escape") {
        e.preventDefault();
        actions.setSearchQuery("");
        inputRef.current?.blur();
      }
    },
    [actions]
  );
  (0, import_react9.useEffect)(() => {
    function handleGlobalKeyDown(e) {
      const mod = e.metaKey || e.ctrlKey;
      if (mod && e.key === "f") {
        e.preventDefault();
        inputRef.current?.focus();
        inputRef.current?.select();
      }
    }
    document.addEventListener("keydown", handleGlobalKeyDown);
    return () => document.removeEventListener("keydown", handleGlobalKeyDown);
  }, []);
  const matchCount = state.searchMatches.length;
  const currentMatch = matchCount > 0 ? state.searchMatchIndex + 1 : 0;
  return /* @__PURE__ */ (0, import_jsx_runtime7.jsxs)(
    "div",
    {
      className,
      style: {
        display: "flex",
        alignItems: "center",
        gap: 6,
        padding: "4px 8px",
        backgroundColor: "var(--vj-bg, #1e1e1e)",
        borderBottom: "1px solid var(--vj-border, #333333)"
      },
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime7.jsx)(
          "input",
          {
            ref: inputRef,
            type: "text",
            value: state.searchQuery,
            onChange: (e) => actions.setSearchQuery(e.target.value),
            onKeyDown: handleKeyDown,
            placeholder: "Search...",
            style: {
              flex: 1,
              background: "none",
              border: "none",
              borderRadius: 3,
              color: "var(--vj-text, #cccccc)",
              fontFamily: "var(--vj-font, monospace)",
              fontSize: "var(--vj-input-font-size, 13px)",
              padding: "3px 8px",
              outline: "none",
              minWidth: 0
            }
          }
        ),
        /* @__PURE__ */ (0, import_jsx_runtime7.jsx)(
          "div",
          {
            style: {
              display: "flex",
              alignItems: "center",
              gap: 2,
              flexShrink: 0,
              height: 18
            },
            children: state.searchQuery ? /* @__PURE__ */ (0, import_jsx_runtime7.jsxs)(import_jsx_runtime7.Fragment, { children: [
              /* @__PURE__ */ (0, import_jsx_runtime7.jsx)(
                "span",
                {
                  style: {
                    fontSize: 11,
                    lineHeight: 1,
                    color: matchCount > 0 ? "var(--vj-text-muted, #999999)" : "var(--vj-error, #f48771)",
                    fontFamily: "var(--vj-font, monospace)",
                    whiteSpace: "nowrap"
                  },
                  children: matchCount > 0 ? `${currentMatch}/${matchCount}` : "0/0"
                }
              ),
              /* @__PURE__ */ (0, import_jsx_runtime7.jsx)(
                "button",
                {
                  onClick: actions.prevSearchMatch,
                  disabled: matchCount === 0,
                  "aria-label": "Previous match (Shift+Enter)",
                  style: {
                    background: "none",
                    border: "none",
                    color: matchCount > 0 ? "var(--vj-text, #cccccc)" : "var(--vj-text-dimmer, #555555)",
                    cursor: matchCount > 0 ? "pointer" : "default",
                    padding: 0,
                    fontSize: 10,
                    lineHeight: 1,
                    display: "inline-flex",
                    alignItems: "center",
                    justifyContent: "center",
                    width: 18,
                    height: 18
                  },
                  title: "Previous match (Shift+Enter)",
                  children: "\u25B2"
                }
              ),
              /* @__PURE__ */ (0, import_jsx_runtime7.jsx)(
                "button",
                {
                  onClick: actions.nextSearchMatch,
                  disabled: matchCount === 0,
                  "aria-label": "Next match (Enter)",
                  style: {
                    background: "none",
                    border: "none",
                    color: matchCount > 0 ? "var(--vj-text, #cccccc)" : "var(--vj-text-dimmer, #555555)",
                    cursor: matchCount > 0 ? "pointer" : "default",
                    padding: 0,
                    fontSize: 10,
                    lineHeight: 1,
                    display: "inline-flex",
                    alignItems: "center",
                    justifyContent: "center",
                    width: 18,
                    height: 18
                  },
                  title: "Next match (Enter)",
                  children: "\u25BC"
                }
              ),
              /* @__PURE__ */ (0, import_jsx_runtime7.jsx)(
                "button",
                {
                  onClick: () => actions.setSearchQuery(""),
                  "aria-label": "Clear search (Esc)",
                  style: {
                    background: "none",
                    border: "none",
                    color: "var(--vj-text, #cccccc)",
                    cursor: "pointer",
                    padding: 0,
                    fontSize: 14,
                    lineHeight: 1,
                    display: "inline-flex",
                    alignItems: "center",
                    justifyContent: "center",
                    width: 18,
                    height: 18
                  },
                  title: "Clear search (Esc)",
                  children: "\xD7"
                }
              )
            ] }) : /* @__PURE__ */ (0, import_jsx_runtime7.jsxs)(import_jsx_runtime7.Fragment, { children: [
              /* @__PURE__ */ (0, import_jsx_runtime7.jsx)(
                "button",
                {
                  onClick: actions.expandAll,
                  "aria-label": "Expand all",
                  style: {
                    background: "none",
                    border: "none",
                    color: "var(--vj-text-muted, #888888)",
                    cursor: "pointer",
                    padding: "2px",
                    fontSize: 12,
                    lineHeight: 1,
                    display: "inline-flex",
                    alignItems: "center"
                  },
                  title: "Expand all",
                  children: /* @__PURE__ */ (0, import_jsx_runtime7.jsxs)(
                    "svg",
                    {
                      width: "14",
                      height: "14",
                      viewBox: "0 0 16 16",
                      fill: "none",
                      stroke: "currentColor",
                      strokeWidth: "1.5",
                      strokeLinecap: "round",
                      strokeLinejoin: "round",
                      children: [
                        /* @__PURE__ */ (0, import_jsx_runtime7.jsx)("path", { d: "M8 2v4M5 4l3-2 3 2" }),
                        /* @__PURE__ */ (0, import_jsx_runtime7.jsx)("path", { d: "M8 14v-4M5 12l3 2 3-2" }),
                        /* @__PURE__ */ (0, import_jsx_runtime7.jsx)("path", { d: "M2 8h12" })
                      ]
                    }
                  )
                }
              ),
              /* @__PURE__ */ (0, import_jsx_runtime7.jsx)(
                "button",
                {
                  onClick: actions.collapseAll,
                  "aria-label": "Collapse all",
                  style: {
                    background: "none",
                    border: "none",
                    color: "var(--vj-text-muted, #888888)",
                    cursor: "pointer",
                    padding: "2px",
                    fontSize: 12,
                    lineHeight: 1,
                    display: "inline-flex",
                    alignItems: "center"
                  },
                  title: "Collapse all",
                  children: /* @__PURE__ */ (0, import_jsx_runtime7.jsxs)(
                    "svg",
                    {
                      width: "14",
                      height: "14",
                      viewBox: "0 0 16 16",
                      fill: "none",
                      stroke: "currentColor",
                      strokeWidth: "1.5",
                      strokeLinecap: "round",
                      strokeLinejoin: "round",
                      children: [
                        /* @__PURE__ */ (0, import_jsx_runtime7.jsx)("path", { d: "M8 5V1M5 3l3 2 3-2" }),
                        /* @__PURE__ */ (0, import_jsx_runtime7.jsx)("path", { d: "M8 11v4M5 13l3-2 3 2" }),
                        /* @__PURE__ */ (0, import_jsx_runtime7.jsx)("path", { d: "M2 8h12" })
                      ]
                    }
                  )
                }
              )
            ] })
          }
        )
      ]
    }
  );
}

// src/json-editor.tsx
var import_jsx_runtime8 = require("react/jsx-runtime");
function JsonEditor({
  value,
  defaultValue,
  onChange,
  schema,
  height = "100%",
  width = "100%",
  className,
  style,
  readOnly,
  treeShowValues = true,
  treeShowCounts = false,
  editorShowDescriptions = false,
  editorShowCounts = false,
  sidebarOpen = true
}) {
  const isControlled = value !== void 0;
  const initialValue = isControlled ? value : defaultValue ?? {};
  const [editorKey, setEditorKey] = (0, import_react10.useState)(0);
  const valueRef = (0, import_react10.useRef)(initialValue);
  (0, import_react10.useEffect)(() => {
    if (isControlled && value !== valueRef.current) {
      valueRef.current = value;
      setEditorKey((k) => k + 1);
    }
  }, [value, isControlled]);
  const handleChange = (0, import_react10.useCallback)(
    (newValue) => {
      valueRef.current = newValue;
      if (!readOnly) {
        onChange?.(newValue);
      }
    },
    [onChange, readOnly]
  );
  const containerStyle = {
    height: typeof height === "number" ? `${height}px` : height,
    width: typeof width === "number" ? `${width}px` : width,
    display: "flex",
    flexDirection: "column",
    overflow: "hidden",
    ...DEFAULT_CSS_VARS,
    ...style
  };
  return /* @__PURE__ */ (0, import_jsx_runtime8.jsxs)("div", { className, "data-vj-root": "", style: containerStyle, children: [
    /* @__PURE__ */ (0, import_jsx_runtime8.jsx)(
      "style",
      {
        dangerouslySetInnerHTML: {
          __html: `@media(pointer:coarse){[data-vj-root]{--vj-input-font-size:16px}}`
        }
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime8.jsx)(
      VisualJson,
      {
        value: valueRef.current,
        onChange: readOnly ? void 0 : handleChange,
        schema,
        children: /* @__PURE__ */ (0, import_jsx_runtime8.jsx)(
          EditorLayout,
          {
            treeShowValues,
            treeShowCounts,
            editorShowDescriptions,
            editorShowCounts,
            sidebarOpen
          }
        )
      },
      editorKey
    )
  ] });
}
function EditorLayout({
  treeShowValues,
  treeShowCounts,
  editorShowDescriptions,
  editorShowCounts,
  sidebarOpen
}) {
  const [sidebarWidth, setSidebarWidth] = (0, import_react10.useState)(280);
  const [isNarrow, setIsNarrow] = (0, import_react10.useState)(false);
  const [activePanel, setActivePanel] = (0, import_react10.useState)("tree");
  const containerRef = (0, import_react10.useRef)(null);
  const dragging = (0, import_react10.useRef)(false);
  const startX = (0, import_react10.useRef)(0);
  const startWidth = (0, import_react10.useRef)(0);
  (0, import_react10.useEffect)(() => {
    function checkWidth() {
      if (containerRef.current) {
        setIsNarrow(containerRef.current.offsetWidth < 500);
      }
    }
    checkWidth();
    const observer = new ResizeObserver(checkWidth);
    if (containerRef.current) observer.observe(containerRef.current);
    return () => observer.disconnect();
  }, []);
  const handleMouseDown = (0, import_react10.useCallback)(
    (e) => {
      dragging.current = true;
      startX.current = e.clientX;
      startWidth.current = sidebarWidth;
      document.body.style.cursor = "col-resize";
      document.body.style.userSelect = "none";
      const handleMouseMove = (ev) => {
        if (!dragging.current) return;
        const delta = ev.clientX - startX.current;
        const newWidth = Math.max(
          180,
          Math.min(600, startWidth.current + delta)
        );
        setSidebarWidth(newWidth);
      };
      const handleMouseUp = () => {
        dragging.current = false;
        document.body.style.cursor = "";
        document.body.style.userSelect = "";
        document.removeEventListener("mousemove", handleMouseMove);
        document.removeEventListener("mouseup", handleMouseUp);
      };
      document.addEventListener("mousemove", handleMouseMove);
      document.addEventListener("mouseup", handleMouseUp);
    },
    [sidebarWidth]
  );
  if (isNarrow) {
    if (!sidebarOpen) {
      return /* @__PURE__ */ (0, import_jsx_runtime8.jsx)(
        "div",
        {
          ref: containerRef,
          style: {
            display: "flex",
            flexDirection: "column",
            flex: 1,
            minHeight: 0
          },
          children: /* @__PURE__ */ (0, import_jsx_runtime8.jsx)("div", { style: { flex: 1, minHeight: 0 }, children: /* @__PURE__ */ (0, import_jsx_runtime8.jsx)(
            FormView,
            {
              showDescriptions: editorShowDescriptions,
              showCounts: editorShowCounts
            }
          ) })
        }
      );
    }
    return /* @__PURE__ */ (0, import_jsx_runtime8.jsxs)(
      "div",
      {
        ref: containerRef,
        style: {
          display: "flex",
          flexDirection: "column",
          flex: 1,
          minHeight: 0
        },
        children: [
          /* @__PURE__ */ (0, import_jsx_runtime8.jsxs)(
            "div",
            {
              style: {
                display: "flex",
                flexShrink: 0,
                borderBottom: "1px solid var(--vj-border, #333333)",
                backgroundColor: "var(--vj-bg-panel, #252526)"
              },
              children: [
                /* @__PURE__ */ (0, import_jsx_runtime8.jsx)(
                  "button",
                  {
                    onClick: () => setActivePanel("tree"),
                    style: {
                      flex: 1,
                      fontSize: 11,
                      padding: "4px 0",
                      cursor: "pointer",
                      fontFamily: "var(--vj-font, monospace)",
                      border: "none",
                      background: activePanel === "tree" ? "var(--vj-accent-muted, #094771)" : "transparent",
                      color: activePanel === "tree" ? "var(--vj-text, #cccccc)" : "var(--vj-text-muted, #999999)"
                    },
                    children: "Tree"
                  }
                ),
                /* @__PURE__ */ (0, import_jsx_runtime8.jsx)(
                  "button",
                  {
                    onClick: () => setActivePanel("form"),
                    style: {
                      flex: 1,
                      fontSize: 11,
                      padding: "4px 0",
                      cursor: "pointer",
                      fontFamily: "var(--vj-font, monospace)",
                      border: "none",
                      background: activePanel === "form" ? "var(--vj-accent-muted, #094771)" : "transparent",
                      color: activePanel === "form" ? "var(--vj-text, #cccccc)" : "var(--vj-text-muted, #999999)"
                    },
                    children: "Form"
                  }
                )
              ]
            }
          ),
          /* @__PURE__ */ (0, import_jsx_runtime8.jsx)("div", { style: { flex: 1, minHeight: 0, overflow: "hidden" }, children: activePanel === "tree" ? /* @__PURE__ */ (0, import_jsx_runtime8.jsxs)(
            "div",
            {
              style: {
                display: "flex",
                flexDirection: "column",
                height: "100%"
              },
              children: [
                /* @__PURE__ */ (0, import_jsx_runtime8.jsx)(SearchBar, {}),
                /* @__PURE__ */ (0, import_jsx_runtime8.jsx)("div", { style: { flex: 1, minHeight: 0, overflow: "auto" }, children: /* @__PURE__ */ (0, import_jsx_runtime8.jsx)(
                  TreeView,
                  {
                    showValues: treeShowValues,
                    showCounts: treeShowCounts
                  }
                ) })
              ]
            }
          ) : /* @__PURE__ */ (0, import_jsx_runtime8.jsx)(
            "div",
            {
              style: {
                display: "flex",
                flexDirection: "column",
                height: "100%"
              },
              children: /* @__PURE__ */ (0, import_jsx_runtime8.jsx)("div", { style: { flex: 1, minHeight: 0 }, children: /* @__PURE__ */ (0, import_jsx_runtime8.jsx)(
                FormView,
                {
                  showDescriptions: editorShowDescriptions,
                  showCounts: editorShowCounts
                }
              ) })
            }
          ) })
        ]
      }
    );
  }
  return /* @__PURE__ */ (0, import_jsx_runtime8.jsxs)("div", { ref: containerRef, style: { display: "flex", flex: 1, minHeight: 0 }, children: [
    /* @__PURE__ */ (0, import_jsx_runtime8.jsxs)(
      "div",
      {
        style: {
          flexShrink: 0,
          overflow: "hidden",
          display: "flex",
          flexDirection: "column",
          width: sidebarOpen ? sidebarWidth : 0,
          transition: "width 0.2s ease"
        },
        children: [
          /* @__PURE__ */ (0, import_jsx_runtime8.jsx)(SearchBar, {}),
          /* @__PURE__ */ (0, import_jsx_runtime8.jsx)("div", { style: { flex: 1, minHeight: 0, overflow: "auto" }, children: /* @__PURE__ */ (0, import_jsx_runtime8.jsx)(TreeView, { showValues: treeShowValues, showCounts: treeShowCounts }) })
        ]
      }
    ),
    sidebarOpen && /* @__PURE__ */ (0, import_jsx_runtime8.jsx)(
      "div",
      {
        style: {
          width: 1,
          flexShrink: 0,
          backgroundColor: "var(--vj-border, #333333)",
          position: "relative",
          transition: "background-color 0.15s"
        },
        children: /* @__PURE__ */ (0, import_jsx_runtime8.jsx)(
          "div",
          {
            onMouseDown: handleMouseDown,
            onMouseEnter: (e) => {
              const parent = e.currentTarget.parentElement;
              if (parent)
                parent.style.backgroundColor = "var(--vj-accent, #007acc)";
            },
            onMouseLeave: (e) => {
              if (!dragging.current) {
                const parent = e.currentTarget.parentElement;
                if (parent)
                  parent.style.backgroundColor = "var(--vj-border, #333333)";
              }
            },
            style: {
              position: "absolute",
              top: 0,
              bottom: 0,
              left: -3,
              width: 7,
              cursor: "col-resize",
              zIndex: 10
            }
          }
        )
      }
    ),
    /* @__PURE__ */ (0, import_jsx_runtime8.jsx)(
      "div",
      {
        style: {
          display: "flex",
          flexDirection: "column",
          flex: 1,
          minWidth: 0,
          overflow: "hidden"
        },
        children: /* @__PURE__ */ (0, import_jsx_runtime8.jsx)("div", { style: { flex: 1, minHeight: 0 }, children: /* @__PURE__ */ (0, import_jsx_runtime8.jsx)(
          FormView,
          {
            showDescriptions: editorShowDescriptions,
            showCounts: editorShowCounts
          }
        ) })
      }
    )
  ] });
}

// src/diff-view.tsx
var import_react11 = require("react");
var import_core8 = require("@visual-json/core");
var import_jsx_runtime9 = require("react/jsx-runtime");
function DiffRow({ entry }) {
  const colors = DIFF_COLORS[entry.type];
  return /* @__PURE__ */ (0, import_jsx_runtime9.jsxs)(
    "div",
    {
      style: {
        display: "flex",
        alignItems: "flex-start",
        padding: "3px 12px",
        borderBottom: "1px solid var(--vj-border-subtle, #2a2a2a)",
        backgroundColor: colors.bg,
        fontFamily: "var(--vj-font, monospace)",
        fontSize: 12,
        gap: 8
      },
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime9.jsx)(
          "span",
          {
            style: {
              color: colors.label,
              fontWeight: 600,
              width: 14,
              flexShrink: 0,
              textAlign: "center"
            },
            children: colors.marker
          }
        ),
        /* @__PURE__ */ (0, import_jsx_runtime9.jsx)(
          "span",
          {
            style: {
              color: "var(--vj-text, #cccccc)",
              flexShrink: 0,
              minWidth: 100
            },
            children: entry.path
          }
        ),
        /* @__PURE__ */ (0, import_jsx_runtime9.jsxs)("span", { style: { flex: 1, display: "flex", gap: 8, overflow: "hidden" }, children: [
          entry.type === "changed" && /* @__PURE__ */ (0, import_jsx_runtime9.jsxs)(import_jsx_runtime9.Fragment, { children: [
            /* @__PURE__ */ (0, import_jsx_runtime9.jsx)("span", { style: { color: "#f48771", textDecoration: "line-through" }, children: formatValue(entry.oldValue) }),
            /* @__PURE__ */ (0, import_jsx_runtime9.jsx)("span", { style: { color: "var(--vj-text-dim, #666666)" }, children: "\u2192" }),
            /* @__PURE__ */ (0, import_jsx_runtime9.jsx)("span", { style: { color: "#4ec94e" }, children: formatValue(entry.newValue) })
          ] }),
          entry.type === "added" && /* @__PURE__ */ (0, import_jsx_runtime9.jsx)("span", { style: { color: "#4ec94e" }, children: formatValue(entry.newValue) }),
          entry.type === "removed" && /* @__PURE__ */ (0, import_jsx_runtime9.jsx)("span", { style: { color: "#f48771", textDecoration: "line-through" }, children: formatValue(entry.oldValue) })
        ] })
      ]
    }
  );
}
function DiffView({
  originalJson,
  currentJson,
  className
}) {
  const entries = (0, import_react11.useMemo)(
    () => (0, import_core8.computeDiff)(originalJson, currentJson),
    [originalJson, currentJson]
  );
  const added = entries.filter((e) => e.type === "added").length;
  const removed = entries.filter((e) => e.type === "removed").length;
  const changed = entries.filter((e) => e.type === "changed").length;
  return /* @__PURE__ */ (0, import_jsx_runtime9.jsxs)(
    "div",
    {
      className,
      style: {
        backgroundColor: "var(--vj-bg, #1e1e1e)",
        color: "var(--vj-text, #cccccc)",
        overflow: "auto",
        height: "100%",
        display: "flex",
        flexDirection: "column"
      },
      children: [
        /* @__PURE__ */ (0, import_jsx_runtime9.jsxs)(
          "div",
          {
            style: {
              display: "flex",
              alignItems: "center",
              gap: 12,
              padding: "6px 12px",
              borderBottom: "1px solid var(--vj-border, #333333)",
              backgroundColor: "var(--vj-bg-panel, #252526)",
              fontFamily: "var(--vj-font, monospace)",
              fontSize: 12,
              flexShrink: 0
            },
            children: [
              /* @__PURE__ */ (0, import_jsx_runtime9.jsx)("span", { style: { color: "var(--vj-text-muted, #999999)" }, children: entries.length === 0 ? "No changes" : `${entries.length} changes` }),
              added > 0 && /* @__PURE__ */ (0, import_jsx_runtime9.jsxs)("span", { style: { color: "#4ec94e" }, children: [
                "+",
                added,
                " added"
              ] }),
              removed > 0 && /* @__PURE__ */ (0, import_jsx_runtime9.jsxs)("span", { style: { color: "#f48771" }, children: [
                "-",
                removed,
                " removed"
              ] }),
              changed > 0 && /* @__PURE__ */ (0, import_jsx_runtime9.jsxs)("span", { style: { color: "#dcdcaa" }, children: [
                "~",
                changed,
                " modified"
              ] })
            ]
          }
        ),
        /* @__PURE__ */ (0, import_jsx_runtime9.jsx)("div", { style: { flex: 1, overflow: "auto" }, children: entries.length === 0 ? /* @__PURE__ */ (0, import_jsx_runtime9.jsx)(
          "div",
          {
            style: {
              display: "flex",
              alignItems: "center",
              justifyContent: "center",
              height: "100%",
              color: "var(--vj-text-dimmer, #555555)",
              fontFamily: "var(--vj-font, monospace)",
              fontSize: 13
            },
            children: "No differences detected"
          }
        ) : entries.map((entry, i) => /* @__PURE__ */ (0, import_jsx_runtime9.jsx)(DiffRow, { entry }, i)) })
      ]
    }
  );
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Breadcrumbs,
  ContextMenu,
  DiffView,
  FormView,
  JsonEditor,
  SearchBar,
  StudioContext,
  TreeView,
  VisualJson,
  getVisibleNodes,
  useStudio
});
